{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "choice-stocks",
   "metadata": {
    "toc": true
   },
   "source": [
    "<h1>Table of Contents<span class=\"tocSkip\"></span></h1>\n",
    "<div class=\"toc\"><ul class=\"toc-item\"><li><span><a href=\"#Loading-Packages\" data-toc-modified-id=\"Loading-Packages-1\"><span class=\"toc-item-num\">1&nbsp;&nbsp;</span>Loading Packages</a></span></li><li><span><a href=\"#Prepping-files-for-SynChro---Finding-Orthologous-pairs\" data-toc-modified-id=\"Prepping-files-for-SynChro---Finding-Orthologous-pairs-2\"><span class=\"toc-item-num\">2&nbsp;&nbsp;</span>Prepping files for SynChro - Finding Orthologous pairs</a></span><ul class=\"toc-item\"><li><span><a href=\"#C.-reinhardtii\" data-toc-modified-id=\"C.-reinhardtii-2.1\"><span class=\"toc-item-num\">2.1&nbsp;&nbsp;</span>C. reinhardtii</a></span></li><li><span><a href=\"#C.-incerta\" data-toc-modified-id=\"C.-incerta-2.2\"><span class=\"toc-item-num\">2.2&nbsp;&nbsp;</span>C. incerta</a></span></li><li><span><a href=\"#C.schloesseri\" data-toc-modified-id=\"C.schloesseri-2.3\"><span class=\"toc-item-num\">2.3&nbsp;&nbsp;</span>C.schloesseri</a></span></li><li><span><a href=\"#JG5-1\" data-toc-modified-id=\"JG5-1-2.4\"><span class=\"toc-item-num\">2.4&nbsp;&nbsp;</span>JG5-1</a></span></li><li><span><a href=\"#C.-malaysian\" data-toc-modified-id=\"C.-malaysian-2.5\"><span class=\"toc-item-num\">2.5&nbsp;&nbsp;</span>C. malaysian</a></span></li></ul></li><li><span><a href=\"#BLASTx---Finding-Orthologous-pairs\" data-toc-modified-id=\"BLASTx---Finding-Orthologous-pairs-3\"><span class=\"toc-item-num\">3&nbsp;&nbsp;</span>BLASTx - Finding Orthologous pairs</a></span><ul class=\"toc-item\"><li><span><a href=\"#Function---Modify-GFF-files\" data-toc-modified-id=\"Function---Modify-GFF-files-3.1\"><span class=\"toc-item-num\">3.1&nbsp;&nbsp;</span>Function - Modify GFF files</a></span></li><li><span><a href=\"#Applying-GFF-function\" data-toc-modified-id=\"Applying-GFF-function-3.2\"><span class=\"toc-item-num\">3.2&nbsp;&nbsp;</span>Applying GFF function</a></span></li><li><span><a href=\"#Getting-longest-transcript/protein\" data-toc-modified-id=\"Getting-longest-transcript/protein-3.3\"><span class=\"toc-item-num\">3.3&nbsp;&nbsp;</span>Getting longest transcript/protein</a></span></li><li><span><a href=\"#Running-BLASTx\" data-toc-modified-id=\"Running-BLASTx-3.4\"><span class=\"toc-item-num\">3.4&nbsp;&nbsp;</span>Running BLASTx</a></span></li></ul></li><li><span><a href=\"#Rob's-Code\" data-toc-modified-id=\"Rob's-Code-4\"><span class=\"toc-item-num\">4&nbsp;&nbsp;</span>Rob's Code</a></span><ul class=\"toc-item\"><li><span><a href=\"#BLAST-line-class\" data-toc-modified-id=\"BLAST-line-class-4.1\"><span class=\"toc-item-num\">4.1&nbsp;&nbsp;</span>BLAST line class</a></span></li><li><span><a href=\"#Assess-the-similarity-of-pairs\" data-toc-modified-id=\"Assess-the-similarity-of-pairs-4.2\"><span class=\"toc-item-num\">4.2&nbsp;&nbsp;</span>Assess the similarity of pairs</a></span></li><li><span><a href=\"#Parse-Reciprocal-Best-BLAST-Hits-(RBH)-into-pairs\" data-toc-modified-id=\"Parse-Reciprocal-Best-BLAST-Hits-(RBH)-into-pairs-4.3\"><span class=\"toc-item-num\">4.3&nbsp;&nbsp;</span>Parse Reciprocal Best BLAST Hits (RBH) into pairs</a></span></li><li><span><a href=\"#Combining-RBH-Orthologous-Pairs\" data-toc-modified-id=\"Combining-RBH-Orthologous-Pairs-4.4\"><span class=\"toc-item-num\">4.4&nbsp;&nbsp;</span>Combining RBH Orthologous Pairs</a></span></li><li><span><a href=\"#Combining-Filtered-RBH-Orthologous-Pairs\" data-toc-modified-id=\"Combining-Filtered-RBH-Orthologous-Pairs-4.5\"><span class=\"toc-item-num\">4.5&nbsp;&nbsp;</span>Combining Filtered RBH Orthologous Pairs</a></span></li><li><span><a href=\"#Sequence-Divergence-of-Orthologous-pairs\" data-toc-modified-id=\"Sequence-Divergence-of-Orthologous-pairs-4.6\"><span class=\"toc-item-num\">4.6&nbsp;&nbsp;</span>Sequence Divergence of Orthologous pairs</a></span></li></ul></li><li><span><a href=\"#Finding-Mating-Type-using-BLASTn\" data-toc-modified-id=\"Finding-Mating-Type-using-BLASTn-5\"><span class=\"toc-item-num\">5&nbsp;&nbsp;</span>Finding Mating Type using BLASTn</a></span></li></ul></div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "contained-angola",
   "metadata": {},
   "source": [
    "# Loading Packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "emotional-cricket",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os, re\n",
    "from itertools import combinations\n",
    "%load_ext rpy2.ipython"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fifty-surveillance",
   "metadata": {},
   "source": [
    "# Prepping files for SynChro - Finding Orthologous pairs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "pressed-shaft",
   "metadata": {},
   "source": [
    "## C. reinhardtii"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "hindu-boards",
   "metadata": {},
   "source": [
    "Please note that in the following code:\n",
    "- Mitochondrion DNA = Chromosome 18 \n",
    "- Chloroplast DNA = Chromosome 19\n",
    "- Mating type locus = Chromosome 20"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "purple-steam",
   "metadata": {},
   "outputs": [],
   "source": [
    "names = ['chr', 'prediction', 'type', 'start', 'end', 'extra1', 'sens', 'extra2', 'extra3']\n",
    "chlam_def = pd.read_csv('/research/references/chlamydomonas/6.1_chlamy/annotation/CreinhardtiiCC_4532_707_v6.1.gene.gff3', delimiter = '\\t', skiprows = 6, names = names)\n",
    "chlam_def['name'] = chlam_def['extra3'].str.split(';', expand = True)[0].str.replace('ID=', '').str.replace('_4532.v6.1', '')\n",
    "chlam_def = chlam_def.drop(columns = ['prediction', 'extra1', 'extra2', 'extra3'])\n",
    "chlam_def.loc[chlam_def.loc[chlam_def['sens'] == '+'].index, 'strand'] = 't'\n",
    "chlam_def.loc[chlam_def.loc[chlam_def['sens'] == '-'].index, 'strand'] = 'f'\n",
    "chlam_def = chlam_def.loc[chlam_def['type'] == 'gene']\n",
    "chlam_def['chr'] = chlam_def['chr'].str.replace('mitogenome', 'chromosome_18').str.replace('plastome', 'chromosome_19').str.replace('mating_type_plus','chromosome_20')\n",
    "chlam_def = chlam_def.loc[chlam_def['chr'].str.contains('chromosome')]\n",
    "chlam_def['chr'] = chlam_def['chr'].str.replace('chromosome_', '')\n",
    "chlam_def[['chr', 'start', 'end']] = chlam_def[['chr', 'start', 'end']].astype(int)\n",
    "chlam_def = chlam_def.sort_values(by = ['chr', 'start'])  ## sorting file according to chromosome number and gene position\n",
    "\n",
    "#### Counting the number of genes per chromosome and per genome ####\n",
    "counter1 = 0 \n",
    "for chrom in chlam_def['chr'].unique():\n",
    "    counter = 0\n",
    "    section = chlam_def.loc[chlam_def['chr'] == chrom]\n",
    "    for idx in section.index:\n",
    "        counter += 1\n",
    "        counter1 += 1\n",
    "        chlam_def.loc[idx, 'IDg/chr'] = counter\n",
    "        chlam_def.loc[idx, 'IDg/all'] = counter1\n",
    "        chlam_def.loc[idx, 'IDf/all'] = counter1\n",
    "        \n",
    "#### Order columns in dataframe ####\n",
    "columns = ['type', 'name', 'chr', 'start', 'end', 'strand', 'sens', 'IDg/chr', 'IDg/all', 'IDf/all']\n",
    "chlam_def = chlam_def[columns]\n",
    "chlam_def[['IDg/chr', 'IDg/all', 'IDf/all']] = chlam_def[['IDg/chr', 'IDg/all', 'IDf/all']].astype(int)\n",
    "\n",
    "#### Save ####\n",
    "chlam_def.to_csv('/research/projects/chlamydomonas/NIexpression/data/CHROnicle/Creinhardtii/01Genomes/CREN.def', sep = '\\t', header = True, index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "spread-absorption",
   "metadata": {},
   "outputs": [],
   "source": [
    "chlam_ch = pd.DataFrame()\n",
    "\n",
    "for i in chlam_def['chr'].unique():\n",
    "    chlam_ch.loc['chromosome', i] = 'chromosome_' + str(i)\n",
    "    chlam_ch.loc['start', i] = chlam_def.loc[chlam_def['chr'] == i]['end'].max()\n",
    "    chlam_ch.loc['end', i] = chlam_def.loc[chlam_def['chr'] == i]['end'].max()\n",
    "\n",
    "#### Save ####\n",
    "chlam_ch.to_csv('/research/projects/chlamydomonas/NIexpression/data/CHROnicle/Creinhardtii/01Genomes/CREN.ch', sep = '\\t', header = False, index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "wireless-delivery",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "\n",
    "original_file = '/research/projects/chlamydomonas/NIexpression/genome_info/ChlreiCC4532_1/CreinhardtiiCC_4532_707_v6.1.protein.fa'\n",
    "corrected_file = \"/research/projects/chlamydomonas/NIexpression/data/CHROnicle/Creinhardtii/01Genomes/CREN.prt\"\n",
    "\n",
    "with open(original_file) as original, open(corrected_file, 'w') as corrected:\n",
    "    records = SeqIO.parse(original_file, 'fasta')\n",
    "    for record in records:\n",
    "        try:\n",
    "            name = record.id.split('_')[0]\n",
    "            idx = chlam_def.index[chlam_def['name'] == name]\n",
    "            a = chlam_def.loc[idx].drop(columns = 'type').values[0].astype(str)\n",
    "            a = '\\t '.join(a)\n",
    "            record.id = a\n",
    "            record.description = a\n",
    "        except:\n",
    "            continue\n",
    "        SeqIO.write(record, corrected, 'fasta')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "outside-static",
   "metadata": {},
   "source": [
    "## C. incerta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "environmental-delta",
   "metadata": {},
   "outputs": [],
   "source": [
    "names = ['chr', 'prediction', 'type', 'start', 'end', 'extra1', 'sens', 'extra2', 'extra3']\n",
    "chlam_def = pd.read_csv('/research/projects/chlamydomonas/Cincerta_deNovo/analysis/assembly_V3/final_assemblies_annotations/Chlamydomonas_incerta/Chlamydomonas_incerta.braker2.gff3', delimiter = '\\t', names = names)\n",
    "chlam_def = chlam_def.loc[chlam_def['type'] == 'gene']\n",
    "chlam_def = chlam_def.sort_values(by = ['chr', 'start'])\n",
    "legend = { chlam_def['chr'].unique()[i]: i + 1 for i in range(len(chlam_def['chr'].unique()))}\n",
    "for i in chlam_def['chr'].unique():\n",
    "    chlam_def.loc[chlam_def.loc[chlam_def['chr'] == i].index, 'chr'] = legend[i]\n",
    "chlam_def.loc[chlam_def.loc[chlam_def['sens'] == '+'].index, 'strand'] = 't'\n",
    "chlam_def.loc[chlam_def.loc[chlam_def['sens'] == '-'].index, 'strand'] = 'f'\n",
    "chlam_def['name'] = chlam_def['extra3'].str.replace('ID=', '').str.replace(';', '')\n",
    "chlam_def = chlam_def.drop(columns = ['prediction', 'extra1', 'extra2', 'extra3'])\n",
    "\n",
    "#### Counting the number of genes per chromosome and per genome ####\n",
    "counter1 = 0 \n",
    "for chrom in chlam_def['chr'].unique():\n",
    "    counter = 0\n",
    "    section = chlam_def.loc[chlam_def['chr'] == chrom]\n",
    "    for idx in section.index:\n",
    "        counter += 1\n",
    "        counter1 += 1\n",
    "        chlam_def.loc[idx, 'IDg/chr'] = counter\n",
    "        chlam_def.loc[idx, 'IDg/all'] = counter1\n",
    "        chlam_def.loc[idx, 'IDf/all'] = counter1\n",
    "        \n",
    "#### Order columns in dataframe ####\n",
    "columns = ['type', 'name', 'chr', 'start', 'end', 'strand', 'sens', 'IDg/chr', 'IDg/all', 'IDf/all']\n",
    "chlam_def = chlam_def[columns]\n",
    "chlam_def[['IDg/chr', 'IDg/all', 'IDf/all']] = chlam_def[['IDg/chr', 'IDg/all', 'IDf/all']].astype(int)\n",
    "\n",
    "#### Save ####\n",
    "chlam_def.to_csv('/research/projects/chlamydomonas/NIexpression/data/CHROnicle/Creinhardtii/01Genomes/CINC.def', sep = '\\t', header = True, index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "premium-ozone",
   "metadata": {},
   "outputs": [],
   "source": [
    "chlam_ch = pd.DataFrame()\n",
    "\n",
    "for i in chlam_def['chr'].unique():\n",
    "    chlam_ch.loc['scaffold', i] = 'scaffold_' + str(i)\n",
    "    chlam_ch.loc['start', i] = chlam_def.loc[chlam_def['chr'] == i]['end'].max()\n",
    "    chlam_ch.loc['end', i] = chlam_def.loc[chlam_def['chr'] == i]['end'].max()\n",
    "\n",
    "#### Save ####\n",
    "chlam_ch.to_csv('/research/projects/chlamydomonas/NIexpression/data/CHROnicle/Creinhardtii/01Genomes/CINC.ch', sep = '\\t', header = False, index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "editorial-modeling",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "\n",
    "original_file = '/research/projects/chlamydomonas/Cincerta_deNovo/analysis/assembly_V3/final_assemblies_annotations/Chlamydomonas_incerta/Chlamydomonas_incerta.braker2.protein.fa'\n",
    "corrected_file = \"/research/projects/chlamydomonas/NIexpression/data/CHROnicle/Creinhardtii/01Genomes/CINC.prt\"\n",
    "\n",
    "with open(original_file) as original, open(corrected_file, 'w') as corrected:\n",
    "    records = SeqIO.parse(original_file, 'fasta')\n",
    "    for record in records:\n",
    "        try:\n",
    "            name = record.id.split('.t')[0]\n",
    "            idx = chlam_def.index[chlam_def['name'] == name]\n",
    "            a = chlam_def.loc[idx].drop(columns = 'type').values[0].astype(str)\n",
    "            a = '\\t '.join(a)\n",
    "            record.id = a\n",
    "            record.description = a\n",
    "        except:\n",
    "            continue\n",
    "        SeqIO.write(record, corrected, 'fasta')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "chicken-contact",
   "metadata": {},
   "source": [
    "## C.schloesseri"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "standing-hypothetical",
   "metadata": {},
   "outputs": [],
   "source": [
    "names = ['chr', 'prediction', 'type', 'start', 'end', 'extra1', 'sens', 'extra2', 'extra3']\n",
    "chlam_def = pd.read_csv('/research/projects/chlamydomonas/Cincerta_deNovo/analysis/assembly_V3/final_assemblies_annotations/Chlamydomonas_schloesseri/Chlamydomonas_schloesseri.braker2.gff3', delimiter = '\\t', names = names)\n",
    "chlam_def = chlam_def.loc[chlam_def['type'] == 'gene']\n",
    "chlam_def = chlam_def.sort_values(by = ['chr', 'start']) \n",
    "legend = { chlam_def['chr'].unique()[i]: i + 1 for i in range(len(chlam_def['chr'].unique()))}\n",
    "for i in chlam_def['chr'].unique():\n",
    "    chlam_def.loc[chlam_def.loc[chlam_def['chr'] == i].index, 'chr'] = legend[i]\n",
    "chlam_def.loc[chlam_def.loc[chlam_def['sens'] == '+'].index, 'strand'] = 't'\n",
    "chlam_def.loc[chlam_def.loc[chlam_def['sens'] == '-'].index, 'strand'] = 'f'\n",
    "chlam_def['name'] = chlam_def['extra3'].str.replace('ID=', '').str.replace(';', '')\n",
    "chlam_def = chlam_def.drop(columns = ['prediction', 'extra1', 'extra2', 'extra3'])\n",
    "\n",
    "#### Counting the number of genes per chromosome and per genome ####\n",
    "counter1 = 0 \n",
    "for chrom in chlam_def['chr'].unique():\n",
    "    counter = 0\n",
    "    section = chlam_def.loc[chlam_def['chr'] == chrom]\n",
    "    for idx in section.index:\n",
    "        counter += 1\n",
    "        counter1 += 1\n",
    "        chlam_def.loc[idx, 'IDg/chr'] = counter\n",
    "        chlam_def.loc[idx, 'IDg/all'] = counter1\n",
    "        chlam_def.loc[idx, 'IDf/all'] = counter1\n",
    "        \n",
    "#### Order columns in dataframe ####\n",
    "columns = ['type', 'name', 'chr', 'start', 'end', 'strand', 'sens', 'IDg/chr', 'IDg/all', 'IDf/all']\n",
    "chlam_def = chlam_def[columns]\n",
    "chlam_def[['IDg/chr', 'IDg/all', 'IDf/all']] = chlam_def[['IDg/chr', 'IDg/all', 'IDf/all']].astype(int)\n",
    "\n",
    "#### Save ####\n",
    "chlam_def.to_csv('/research/projects/chlamydomonas/NIexpression/data/CHROnicle/Creinhardtii/01Genomes/CSCH.def', sep = '\\t', header = True, index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "isolated-sharing",
   "metadata": {},
   "outputs": [],
   "source": [
    "chlam_ch = pd.DataFrame()\n",
    "\n",
    "for i in chlam_def['chr'].unique():\n",
    "    chlam_ch.loc['scaffold', i] = 'scaffold_' + str(i)\n",
    "    chlam_ch.loc['start', i] = chlam_def.loc[chlam_def['chr'] == i]['end'].max()\n",
    "    chlam_ch.loc['end', i] = chlam_def.loc[chlam_def['chr'] == i]['end'].max()\n",
    "\n",
    "#### Save ####\n",
    "chlam_ch.to_csv('/research/projects/chlamydomonas/NIexpression/data/CHROnicle/Creinhardtii/01Genomes/CSCH.ch', sep = '\\t', header = False, index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "behavioral-whole",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "\n",
    "original_file = '/research/projects/chlamydomonas/Cincerta_deNovo/analysis/assembly_V3/final_assemblies_annotations/Chlamydomonas_schloesseri/Chlamydomonas_schloesseri.braker2.protein.fa'\n",
    "corrected_file = \"/research/projects/chlamydomonas/NIexpression/data/CHROnicle/Creinhardtii/01Genomes/CSCH.prt\"\n",
    "\n",
    "with open(original_file) as original, open(corrected_file, 'w') as corrected:\n",
    "    records = SeqIO.parse(original_file, 'fasta')\n",
    "    for record in records:\n",
    "        try:\n",
    "            name = record.id.split('.t')[0]\n",
    "            idx = chlam_def.index[chlam_def['name'] == name]\n",
    "            a = chlam_def.loc[idx].drop(columns = 'type').values[0].astype(str)\n",
    "            a = '\\t '.join(a)\n",
    "            record.id = a\n",
    "            record.description = a\n",
    "        except:\n",
    "            continue\n",
    "        SeqIO.write(record, corrected, 'fasta')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "hollow-amendment",
   "metadata": {},
   "source": [
    "## JG5-1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "geographic-multiple",
   "metadata": {},
   "outputs": [],
   "source": [
    "name = ['chr', 'prediction', 'type', 'start', 'end', 'extra1', 'sens', 'extra2', 'extra3']\n",
    "chlam_def = pd.read_csv('/research/references/chlamydomonas/JG5.v1/JG5.v1.genome_w_organelles.gtf', delimiter = '\\t', names = name, skiprows = 1)\n",
    "chlam_def = chlam_def.loc[chlam_def['type'] == 'gene']\n",
    "chlam_def['name'] = chlam_def['extra3'].str.split(';', expand = True)[0].str.replace('ID=', '')\n",
    "chlam_def = chlam_def.drop(columns = ['prediction', 'extra1', 'extra2', 'extra3'])\n",
    "chlam_def.loc[chlam_def.loc[chlam_def['sens'] == '+'].index, 'strand'] = 't'\n",
    "chlam_def.loc[chlam_def.loc[chlam_def['sens'] == '-'].index, 'strand'] = 'f'\n",
    "chlam_def['chr'] = chlam_def['chr'].str.replace('JG5-1_mitogenome', '91').str.replace('JG5-1_plastogenome', '92').str.replace('mating_type_plus','chromosome_20')\n",
    "\n",
    "legend = {chlam_def['chr'].unique()[i + 2]: i+1 for i in range(90)}\n",
    "for i in legend.keys():\n",
    "    chlam_def.loc[chlam_def.loc[chlam_def['chr'] == i].index, 'chr'] = legend[i]\n",
    "    \n",
    "chlam_def[['chr', 'start', 'end']] = chlam_def[['chr', 'start', 'end']].astype(int)\n",
    "chlam_def = chlam_def.sort_values(by = ['chr', 'start'])  ## sorting file according to chromosome number and gene position\n",
    "\n",
    "#### Assigning chromosome number based on contigs ####\n",
    "counter = 0\n",
    "for i in chlam_def['chr'].unique():\n",
    "    counter += 1\n",
    "    chlam_def.loc[chlam_def.loc[chlam_def['chr'] == i].index, 'chr'] = counter\n",
    "\n",
    "#### Counting the number of genes per chromosome and per genome ####\n",
    "counter1 = 0 \n",
    "for chrom in chlam_def['chr'].unique():\n",
    "    counter = 0\n",
    "    section = chlam_def.loc[chlam_def['chr'] == chrom]\n",
    "    for idx in section.index:\n",
    "        counter += 1\n",
    "        counter1 += 1\n",
    "        chlam_def.loc[idx, 'IDg/chr'] = counter\n",
    "        chlam_def.loc[idx, 'IDg/all'] = counter1\n",
    "        chlam_def.loc[idx, 'IDf/all'] = counter1\n",
    "        \n",
    "#### Order columns in dataframe ####\n",
    "columns = ['type', 'name', 'chr', 'start', 'end', 'strand', 'sens', 'IDg/chr', 'IDg/all', 'IDf/all']\n",
    "chlam_def = chlam_def[columns]\n",
    "chlam_def[['IDg/chr', 'IDg/all', 'IDf/all']] = chlam_def[['IDg/chr', 'IDg/all', 'IDf/all']].astype(int)\n",
    "\n",
    "### Save ####\n",
    "chlam_def.to_csv('/research/projects/chlamydomonas/NIexpression/data/CHROnicle/Creinhardtii/01Genomes/JG05.def', sep = '\\t', header = True, index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "black-catalog",
   "metadata": {},
   "outputs": [],
   "source": [
    "chlam_ch = pd.DataFrame()\n",
    "\n",
    "for i in chlam_def['chr'].unique():\n",
    "    chlam_ch.loc['chr', i] = 'scaffold_' + str(i)\n",
    "    chlam_ch.loc['start', i] = chlam_def.loc[chlam_def['chr'] == i]['end'].max()\n",
    "    chlam_ch.loc['end', i] = chlam_def.loc[chlam_def['chr'] == i]['end'].max()\n",
    "\n",
    "#### Save ####\n",
    "chlam_ch.to_csv('/research/projects/chlamydomonas/NIexpression/data/CHROnicle/Creinhardtii/01Genomes/JG05.ch', sep = '\\t', header = False, index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "legislative-delivery",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "\n",
    "original_file = '/research/projects/chlamydomonas/NIexpression/genome_info/JG5/JG5-1_protein.fasta'\n",
    "corrected_file = \"/research/projects/chlamydomonas/NIexpression/data/CHROnicle/Creinhardtii/01Genomes/JG05.prt\"\n",
    "\n",
    "with open(original_file) as original, open(corrected_file, 'w') as corrected:\n",
    "    records = SeqIO.parse(original_file, 'fasta')\n",
    "    for record in records:\n",
    "        try:\n",
    "            name = record.id.replace('.t1', '')\n",
    "            idx = chlam_def.index[chlam_def['name'] == name]\n",
    "            a = chlam_def.loc[idx].drop(columns = 'type').values[0].astype(str)\n",
    "            a = '\\t '.join(a)\n",
    "            record.id = a\n",
    "            record.description = a\n",
    "        except:\n",
    "            try:\n",
    "                name = record.id.replace('cds', 'gene')\n",
    "                idx = chlam_def.index[chlam_def['name'] == name]\n",
    "                a = chlam_def.loc[idx].drop(columns = 'type').values[0].astype(str)\n",
    "                a = '\\t '.join(a)\n",
    "                record.id = a\n",
    "                record.description = a\n",
    "            except:\n",
    "                continue \n",
    "        SeqIO.write(record, corrected, 'fasta')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "basic-addition",
   "metadata": {},
   "source": [
    "## C. malaysian"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "imported-terrorism",
   "metadata": {},
   "outputs": [],
   "source": [
    "name = ['chr', 'prediction', 'type', 'start', 'end', 'extra1', 'sens', 'extra2', 'extra3']\n",
    "chlam_def = pd.read_csv('/research/references/chlamydomonas/Cmalay.v1/Cmalay.v1.genome_w_organelles.gtf', delimiter = '\\t', names = name, skiprows = 1)\n",
    "chlam_def = chlam_def.loc[chlam_def['type'] == 'gene']\n",
    "chlam_def['name'] = chlam_def['extra3'].str.split(';', expand = True)[0].str.replace('ID=', '')\n",
    "chlam_def = chlam_def.drop(columns = ['prediction', 'extra1', 'extra2', 'extra3'])\n",
    "chlam_def.loc[chlam_def.loc[chlam_def['sens'] == '+'].index, 'strand'] = 't'\n",
    "chlam_def.loc[chlam_def.loc[chlam_def['sens'] == '-'].index, 'strand'] = 'f'\n",
    "chlam_def['chr'] = chlam_def['chr'].str.replace('C_malay_mitogenome', '63').str.replace('C_malay_plastogenome', '64').str.replace('mating_type_plus','chromosome_20')\n",
    "\n",
    "legend = {chlam_def['chr'].unique()[i + 2]: i+1 for i in range(62)}\n",
    "for i in legend.keys():\n",
    "    chlam_def.loc[chlam_def.loc[chlam_def['chr'] == i].index, 'chr'] = legend[i]\n",
    "    \n",
    "chlam_def[['chr', 'start', 'end']] = chlam_def[['chr', 'start', 'end']].astype(int)\n",
    "chlam_def = chlam_def.sort_values(by = ['chr', 'start'])  ## sorting file according to chromosome number and gene position\n",
    "\n",
    "#### Assigning chromosome number based on contigs ####\n",
    "counter = 0\n",
    "for i in chlam_def['chr'].unique():\n",
    "    counter += 1\n",
    "    chlam_def.loc[chlam_def.loc[chlam_def['chr'] == i].index, 'chr'] = counter\n",
    "\n",
    "#### Counting the number of genes per chromosome and per genome ####\n",
    "counter1 = 0 \n",
    "for chrom in chlam_def['chr'].unique():\n",
    "    counter = 0\n",
    "    section = chlam_def.loc[chlam_def['chr'] == chrom]\n",
    "    for idx in section.index:\n",
    "        counter += 1\n",
    "        counter1 += 1\n",
    "        chlam_def.loc[idx, 'IDg/chr'] = counter\n",
    "        chlam_def.loc[idx, 'IDg/all'] = counter1\n",
    "        chlam_def.loc[idx, 'IDf/all'] = counter1\n",
    "        \n",
    "#### Order columns in dataframe ####\n",
    "columns = ['type', 'name', 'chr', 'start', 'end', 'strand', 'sens', 'IDg/chr', 'IDg/all', 'IDf/all']\n",
    "chlam_def = chlam_def[columns]\n",
    "chlam_def[['IDg/chr', 'IDg/all', 'IDf/all']] = chlam_def[['IDg/chr', 'IDg/all', 'IDf/all']].astype(int)\n",
    "\n",
    "### Save ####\n",
    "chlam_def.to_csv('/research/projects/chlamydomonas/NIexpression/data/CHROnicle/Creinhardtii/01Genomes/MALA.def', sep = '\\t', header = True, index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "divine-nutrition",
   "metadata": {},
   "outputs": [],
   "source": [
    "chlam_ch = pd.DataFrame()\n",
    "\n",
    "for i in chlam_def['chr'].unique():\n",
    "    chlam_ch.loc['chr', i] = 'scaffold_' + str(i)\n",
    "    chlam_ch.loc['start', i] = chlam_def.loc[chlam_def['chr'] == i]['end'].max()\n",
    "    chlam_ch.loc['end', i] = chlam_def.loc[chlam_def['chr'] == i]['end'].max()\n",
    "\n",
    "#### Save ####\n",
    "chlam_ch.to_csv('/research/projects/chlamydomonas/NIexpression/data/CHROnicle/Creinhardtii/01Genomes/MALA.ch', sep = '\\t', header = False, index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "declared-validity",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "\n",
    "original_file = '/research/projects/chlamydomonas/NIexpression/genome_info/Malaysian/C_malay_protein.fasta'\n",
    "corrected_file = \"/research/projects/chlamydomonas/NIexpression/data/CHROnicle/Creinhardtii/01Genomes/MALA.prt\"\n",
    "\n",
    "with open(original_file) as original, open(corrected_file, 'w') as corrected:\n",
    "    records = SeqIO.parse(original_file, 'fasta')\n",
    "    for record in records:\n",
    "        try:\n",
    "            name = record.id.replace('.t1', '')\n",
    "            idx = chlam_def.index[chlam_def['name'] == name]\n",
    "            a = chlam_def.loc[idx].drop(columns = 'type').values[0].astype(str)\n",
    "            a = '\\t '.join(a)\n",
    "            record.id = a\n",
    "            record.description = a\n",
    "        except:\n",
    "            try:\n",
    "                name = record.id.replace('cds', 'gene')\n",
    "                idx = chlam_def.index[chlam_def['name'] == name]\n",
    "                a = chlam_def.loc[idx].drop(columns = 'type').values[0].astype(str)\n",
    "                a = '\\t '.join(a)\n",
    "                record.id = a\n",
    "                record.description = a\n",
    "            except:\n",
    "                continue \n",
    "        SeqIO.write(record, corrected, 'fasta')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "pending-harvey",
   "metadata": {},
   "source": [
    "# BLASTx - Finding Orthologous pairs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "beautiful-microphone",
   "metadata": {},
   "source": [
    "## Function - Modify GFF files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "successful-auction",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "def modified_gff(gff):\n",
    "    ''' This code is meant to remove all duplicated transcripts and leave behind the longest protein isoform'''\n",
    "    mRNA = gff.loc[(gff['type'] == 'mRNA') | (gff['type'] == 'transcript')]\n",
    "    mRNA['attributes'].str.split(';', expand = True)\n",
    "    mRNA[['ID', 'Parent']] = mRNA['attributes'].str.split(';', expand = True)[[0, 1]]\n",
    "    mRNA['genelength'] = mRNA['end'] - mRNA['start']\n",
    "    \n",
    "    #### Drop transcripts with same length and lowest score ####\n",
    "    for parent in mRNA.loc[mRNA.duplicated(subset = ['Parent'])]['Parent'].unique():\n",
    "        parent_section = mRNA.loc[mRNA['Parent'] == parent]\n",
    "        max_score = parent_section['score'].max()\n",
    "        mRNA.drop(parent_section.loc[parent_section['score'] != max_score].index, inplace = True)\n",
    "    \n",
    "    #### Remove shortest transcripts ####\n",
    "    for i in mRNA['Parent'][mRNA.duplicated(subset=['Parent'])]:\n",
    "        section = mRNA.loc[mRNA['Parent'] == i]\n",
    "        mx = mRNA.loc[mRNA['Parent'] == i]['genelength'].max()\n",
    "        mRNA.drop(section.loc[section['genelength'] != mx].index, inplace = True)\n",
    "        \n",
    "    mRNA.drop_duplicates(subset = ['Parent', 'genelength'], keep = 'first', inplace = True)\n",
    "        \n",
    "    mRNA['ID'] = mRNA['ID'].str.replace('ID=', '')\n",
    "    mRNA['ID'] = mRNA['ID'].str.replace('transcript_id', '').str.replace('\"', '').str.replace(' ', '')\n",
    "    return mRNA"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "through-lighting",
   "metadata": {},
   "source": [
    "## Applying GFF function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 182,
   "id": "public-teach",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-91-8a38c5fa11a7>:5: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  mRNA[['ID', 'Parent']] = mRNA['attributes'].str.split(';', expand = True)[[0, 1]]\n",
      "<ipython-input-91-8a38c5fa11a7>:5: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  mRNA[['ID', 'Parent']] = mRNA['attributes'].str.split(';', expand = True)[[0, 1]]\n",
      "<ipython-input-91-8a38c5fa11a7>:6: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  mRNA['genelength'] = mRNA['end'] - mRNA['start']\n",
      "<ipython-input-91-8a38c5fa11a7>:12: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  mRNA.drop(parent_section.loc[parent_section['score'] != max_score].index, inplace = True)\n",
      "<ipython-input-91-8a38c5fa11a7>:18: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  mRNA.drop(section.loc[section['genelength'] != mx].index, inplace = True)\n",
      "<ipython-input-91-8a38c5fa11a7>:20: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  mRNA.drop_duplicates(subset = ['Parent', 'genelength'], keep = 'first', inplace = True)\n",
      "<ipython-input-91-8a38c5fa11a7>:22: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  mRNA['ID'] = mRNA['ID'].str.replace('ID=', '')\n",
      "<ipython-input-91-8a38c5fa11a7>:23: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  mRNA['ID'] = mRNA['ID'].str.replace('transcript_id', '').str.replace('\"', '').str.replace(' ', '')\n",
      "<ipython-input-91-8a38c5fa11a7>:5: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  mRNA[['ID', 'Parent']] = mRNA['attributes'].str.split(';', expand = True)[[0, 1]]\n",
      "<ipython-input-91-8a38c5fa11a7>:5: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  mRNA[['ID', 'Parent']] = mRNA['attributes'].str.split(';', expand = True)[[0, 1]]\n",
      "<ipython-input-91-8a38c5fa11a7>:6: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  mRNA['genelength'] = mRNA['end'] - mRNA['start']\n",
      "<ipython-input-91-8a38c5fa11a7>:12: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  mRNA.drop(parent_section.loc[parent_section['score'] != max_score].index, inplace = True)\n",
      "<ipython-input-91-8a38c5fa11a7>:18: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  mRNA.drop(section.loc[section['genelength'] != mx].index, inplace = True)\n",
      "<ipython-input-91-8a38c5fa11a7>:20: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  mRNA.drop_duplicates(subset = ['Parent', 'genelength'], keep = 'first', inplace = True)\n",
      "<ipython-input-91-8a38c5fa11a7>:22: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  mRNA['ID'] = mRNA['ID'].str.replace('ID=', '')\n",
      "<ipython-input-91-8a38c5fa11a7>:23: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  mRNA['ID'] = mRNA['ID'].str.replace('transcript_id', '').str.replace('\"', '').str.replace(' ', '')\n",
      "<ipython-input-91-8a38c5fa11a7>:5: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  mRNA[['ID', 'Parent']] = mRNA['attributes'].str.split(';', expand = True)[[0, 1]]\n",
      "<ipython-input-91-8a38c5fa11a7>:5: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  mRNA[['ID', 'Parent']] = mRNA['attributes'].str.split(';', expand = True)[[0, 1]]\n",
      "<ipython-input-91-8a38c5fa11a7>:6: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  mRNA['genelength'] = mRNA['end'] - mRNA['start']\n",
      "<ipython-input-91-8a38c5fa11a7>:12: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  mRNA.drop(parent_section.loc[parent_section['score'] != max_score].index, inplace = True)\n",
      "<ipython-input-91-8a38c5fa11a7>:18: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  mRNA.drop(section.loc[section['genelength'] != mx].index, inplace = True)\n",
      "<ipython-input-91-8a38c5fa11a7>:20: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  mRNA.drop_duplicates(subset = ['Parent', 'genelength'], keep = 'first', inplace = True)\n",
      "<ipython-input-91-8a38c5fa11a7>:22: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  mRNA['ID'] = mRNA['ID'].str.replace('ID=', '')\n",
      "<ipython-input-91-8a38c5fa11a7>:23: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  mRNA['ID'] = mRNA['ID'].str.replace('transcript_id', '').str.replace('\"', '').str.replace(' ', '')\n",
      "<ipython-input-91-8a38c5fa11a7>:5: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  mRNA[['ID', 'Parent']] = mRNA['attributes'].str.split(';', expand = True)[[0, 1]]\n",
      "<ipython-input-91-8a38c5fa11a7>:5: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  mRNA[['ID', 'Parent']] = mRNA['attributes'].str.split(';', expand = True)[[0, 1]]\n",
      "<ipython-input-91-8a38c5fa11a7>:6: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  mRNA['genelength'] = mRNA['end'] - mRNA['start']\n",
      "<ipython-input-91-8a38c5fa11a7>:12: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  mRNA.drop(parent_section.loc[parent_section['score'] != max_score].index, inplace = True)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-91-8a38c5fa11a7>:18: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  mRNA.drop(section.loc[section['genelength'] != mx].index, inplace = True)\n",
      "<ipython-input-91-8a38c5fa11a7>:20: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  mRNA.drop_duplicates(subset = ['Parent', 'genelength'], keep = 'first', inplace = True)\n",
      "<ipython-input-91-8a38c5fa11a7>:22: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  mRNA['ID'] = mRNA['ID'].str.replace('ID=', '')\n",
      "<ipython-input-91-8a38c5fa11a7>:23: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  mRNA['ID'] = mRNA['ID'].str.replace('transcript_id', '').str.replace('\"', '').str.replace(' ', '')\n",
      "<ipython-input-182-c01a08f3cf8e>:13: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  JG5_gff2['Parent'] = JG5_gff2['Parent'].str.replace('gene_id', '').str.replace('\"', '').str.replace(' ', '')\n",
      "<ipython-input-182-c01a08f3cf8e>:14: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  Csch_gff2['Parent'] = Csch_gff2['Parent'].str.replace('Parent=', '').str.replace(' ', '')\n",
      "<ipython-input-182-c01a08f3cf8e>:15: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  Cinc_gff2['Parent'] = Cinc_gff2['Parent'].str.replace('Parent=', '').str.replace(' ', '')\n",
      "<ipython-input-182-c01a08f3cf8e>:16: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  Cmal_gff2['Parent'] = Cmal_gff2['Parent'].str.replace('gene_id', '').str.replace('\"', '').str.replace(' ', '')\n"
     ]
    }
   ],
   "source": [
    "name = ['seqid', 'source', 'type', 'start', 'end', 'score', 'strand', 'phase', 'attributes'] \n",
    "\n",
    "Cinc_gff = pd.read_csv('/research/projects/chlamydomonas/Cincerta_deNovo/analysis/assembly_V3/final_assemblies_annotations/Chlamydomonas_incerta/Chlamydomonas_incerta.braker2.gff3', delimiter = '\\t', names = name)\n",
    "Csch_gff = pd.read_csv('/research/projects/chlamydomonas/Cincerta_deNovo/analysis/assembly_V3/final_assemblies_annotations/Chlamydomonas_schloesseri/Chlamydomonas_schloesseri.braker2.gff3', delimiter = '\\t', names = name)\n",
    "JG5_gff = pd.read_csv('/research/projects/chlamydomonas/NIexpression/genome_info/JG5/JG5-1.v1.genome_w_organelles.gtf', delimiter = '\\t', names = name)\n",
    "Cmal_gff = pd.read_csv('/research/projects/chlamydomonas/NIexpression/genome_info/Malaysian/Cmalay.v1.genome_w_organelles.gtf', delimiter = '\\t', names = name)\n",
    "\n",
    "Cinc_gff2 = modified_gff(Cinc_gff)\n",
    "Csch_gff2 = modified_gff(Csch_gff)\n",
    "JG5_gff2 = modified_gff(JG5_gff)\n",
    "Cmal_gff2 = modified_gff(Cmal_gff)\n",
    "\n",
    "JG5_gff2['Parent'] = JG5_gff2['Parent'].str.replace('gene_id', '').str.replace('\"', '').str.replace(' ', '')\n",
    "Csch_gff2['Parent'] = Csch_gff2['Parent'].str.replace('Parent=', '').str.replace(' ', '')\n",
    "Cinc_gff2['Parent'] = Cinc_gff2['Parent'].str.replace('Parent=', '').str.replace(' ', '')\n",
    "Cmal_gff2['Parent'] = Cmal_gff2['Parent'].str.replace('gene_id', '').str.replace('\"', '').str.replace(' ', '')\n",
    "\n",
    "Cren_gff2 = pd.read_csv('/research/references/chlamydomonas/6.1_chlamy/annotation/CreinhardtiiCC_4532_707_v6.1.gene.gff3', delimiter = '\\t', skiprows = 6, names = name)\n",
    "Cren_gff2['attributes'] = Cren_gff2['attributes'].str.replace('ID=', '').str.split('_4532', expand = True)[0]\n",
    "Cren_gff2 = Cren_gff2.loc[Cren_gff2['type'] == 'mRNA']\n",
    "Cren_gff2['readlength'] = Cren_gff2['end'] - Cren_gff2['start']\n",
    "Cren_gff2.drop_duplicates(subset = ['seqid', 'attributes', 'readlength'], keep = 'first', inplace = True)\n",
    "\n",
    "#### GFF3 ####\n",
    "for transcript in Cren_gff2['attributes'].unique():\n",
    "    section = Cren_gff2.loc[Cren_gff2['attributes'] == transcript]\n",
    "    transcript_max = section['readlength'].max()\n",
    "    Cren_gff2.drop(section.loc[section['readlength'] != transcript_max].index, axis = 0, inplace = True)\n",
    "Cren_gff2.columns = Cren_gff2.columns.str.replace('attributes', 'Parent')\n",
    "Cren_gff2.sort_values(by = ['seqid', 'start'], ascending = True, inplace = True)\n",
    "\n",
    "#### Save new GFF3 file ####\n",
    "JG5_gff2.to_csv('/research/projects/chlamydomonas/NIexpression/genome_info/JG5/JG5_transcript.gff3', sep = '\\t', index = False, header = True)\n",
    "Csch_gff2.to_csv('/research/projects/chlamydomonas/NIexpression/genome_info/Cschloesseri/Cschloesseri_mRNA.gff3', sep = '\\t', index = False, header = True)\n",
    "Cinc_gff2.to_csv('/research/projects/chlamydomonas/NIexpression/genome_info/Cincerta/Cincerta_mRNA.gff3', sep = '\\t', index = False, header = True)\n",
    "Cmal_gff2.to_csv('/research/projects/chlamydomonas/NIexpression/genome_info/Malaysian/Malaysian_transcript.gff3', sep = '\\t', index = False, header = True)\n",
    "Cren_gff2.to_csv('/research/projects/chlamydomonas/NIexpression/genome_info/Creinhardtii_v6.1/Creinhardtii_mRNA.gff3', sep = '\\t', index = False, header = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "demanding-transportation",
   "metadata": {},
   "source": [
    "## Getting longest transcript/protein"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "incorporated-portal",
   "metadata": {},
   "source": [
    "The code below provides a fasta file with the longest transcript per gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 165,
   "id": "scenic-alberta",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "\n",
    "Cinc_fasta = \"/research/projects/chlamydomonas/Cincerta_deNovo/analysis/assembly_V3/final_assemblies_annotations/Chlamydomonas_incerta/Chlamydomonas_incerta.braker2.CDS.fa\"\n",
    "Csch_fasta = '/research/projects/chlamydomonas/Cincerta_deNovo/analysis/assembly_V3/final_assemblies_annotations/Chlamydomonas_schloesseri/Chlamydomonas_schloesseri.braker2.CDS.fa'\n",
    "Cmal_fasta = '/research/projects/chlamydomonas/NIexpression/genome_info/Malaysian/Cmalay.CDS.fa'\n",
    "JG5_fasta = '/research/projects/chlamydomonas/NIexpression/genome_info/JG5/JG5.CDS.fa'\n",
    "\n",
    "Cinc_prot = '/research/projects/chlamydomonas/Cincerta_deNovo/analysis/assembly_V3/final_assemblies_annotations/Chlamydomonas_incerta/Chlamydomonas_incerta.braker2.protein.fa'\n",
    "Csch_prot = '/research/projects/chlamydomonas/Cincerta_deNovo/analysis/assembly_V3/final_assemblies_annotations/Chlamydomonas_schloesseri/Chlamydomonas_schloesseri.braker2.protein.fa'\n",
    "\n",
    "############################\n",
    "#### Longest Transcript ####\n",
    "############################\n",
    "cinc_refn = SeqIO.to_dict(SeqIO.parse(open(Cinc_fasta), 'fasta'))\n",
    "o=open(\"/research/projects/chlamydomonas/NIexpression/genome_info/Cincerta/longest_isoform_transcript.fa\", 'w')\n",
    "for t_id in Cinc_gff2['ID'].unique():\n",
    "    cds = \"\".join(cinc_refn[t_id])\n",
    "    o.write(\">{t_id}|Cincerta\\n{cds}\\n\".format(t_id=t_id,cds=cds) )\n",
    "o.close()\n",
    "\n",
    "csch_refn = SeqIO.to_dict(SeqIO.parse(open(Csch_fasta), 'fasta'))\n",
    "o=open(\"/research/projects/chlamydomonas/NIexpression/genome_info/Cschloesseri/longest_isoform_transcript.fa\", 'w')\n",
    "for t_id in Csch_gff2['ID'].unique():\n",
    "    cds = \"\".join(csch_refn[t_id])\n",
    "    o.write(\">{t_id}|Cschloesseri\\n{cds}\\n\".format(t_id=t_id,cds=cds) )\n",
    "o.close()\n",
    "\n",
    "cmal_refn = SeqIO.to_dict(SeqIO.parse(open(Cmal_fasta), 'fasta'))\n",
    "o=open(\"/research/projects/chlamydomonas/NIexpression/genome_info/Malaysian/longest_isoform_transcript.fa\", 'w')\n",
    "for t_id in Cmal_gff2['ID'].unique():\n",
    "    try:\n",
    "        cds = \"\".join(cmal_refn[t_id])\n",
    "        o.write(\">{t_id}|Cmalaysian\\n{cds}\\n\".format(t_id=t_id,cds=cds) )\n",
    "    except: continue\n",
    "o.close()\n",
    "\n",
    "jg5_refn = SeqIO.to_dict(SeqIO.parse(open(JG5_fasta), 'fasta'))\n",
    "o=open(\"/research/projects/chlamydomonas/NIexpression/genome_info/JG5/longest_isoform_transcript.fa\", 'w')\n",
    "for t_id in JG5_gff2['ID'].unique():\n",
    "    try:\n",
    "        cds = \"\".join(jg5_refn[t_id])\n",
    "        o.write(\">{t_id}|JG5\\n{cds}\\n\".format(t_id=t_id,cds=cds) )\n",
    "    except: continue\n",
    "o.close()\n",
    "\n",
    "#################################\n",
    "#### Longest Protein Isoform ####\n",
    "#################################\n",
    "cinc_refp = SeqIO.to_dict(SeqIO.parse(open(Cinc_prot), 'fasta'))\n",
    "o=open(\"/research/projects/chlamydomonas/NIexpression/genome_info/Cincerta/longest_isoform_protein.fa\", 'w')\n",
    "for t_id in Cinc_gff2['ID'].unique():\n",
    "    cds = \"\".join(cinc_refp[t_id])\n",
    "    o.write(\">{t_id}|Cincerta\\n{cds}\\n\".format(t_id=t_id,cds=cds) )\n",
    "o.close()\n",
    "\n",
    "csch_refp = SeqIO.to_dict(SeqIO.parse(open(Csch_prot), 'fasta'))\n",
    "o=open(\"/research/projects/chlamydomonas/NIexpression/genome_info/Cschloesseri/longest_isoform_protein.fa\", 'w')\n",
    "for t_id in Csch_gff2['ID'].unique():\n",
    "    cds = \"\".join(csch_refp[t_id])\n",
    "    o.write(\">{t_id}|Cschloesseri\\n{cds}\\n\".format(t_id=t_id,cds=cds) )\n",
    "o.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "falling-tender",
   "metadata": {},
   "source": [
    "## Running BLASTx "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 166,
   "id": "mineral-disclosure",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "Building a new DB, current time: 05/04/2025 19:05:09\n",
      "New DB name:   /research/projects/chlamydomonas/NIexpression/genome_info/Creinhardtii_v6.1/CreinhardtiiCC_4532_707_v6.1.transcript_primaryTranscriptOnly.fa\n",
      "New DB title:  /research/projects/chlamydomonas/NIexpression/genome_info/Creinhardtii_v6.1/CreinhardtiiCC_4532_707_v6.1.transcript_primaryTranscriptOnly.fa\n",
      "Sequence type: Nucleotide\n",
      "Deleted existing Nucleotide BLAST database named /research/projects/chlamydomonas/NIexpression/genome_info/Creinhardtii_v6.1/CreinhardtiiCC_4532_707_v6.1.transcript_primaryTranscriptOnly.fa\n",
      "Keep MBits: T\n",
      "Maximum file size: 3000000000B\n",
      "Adding sequences from FASTA; added 17693 sequences in 1.18837 seconds.\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Building a new DB, current time: 05/04/2025 19:05:10\n",
      "New DB name:   /research/projects/chlamydomonas/NIexpression/genome_info/Cincerta/longest_isoform_transcript.fa\n",
      "New DB title:  /research/projects/chlamydomonas/NIexpression/genome_info/Cincerta/longest_isoform_transcript.fa\n",
      "Sequence type: Nucleotide\n",
      "Deleted existing Nucleotide BLAST database named /research/projects/chlamydomonas/NIexpression/genome_info/Cincerta/longest_isoform_transcript.fa\n",
      "Keep MBits: T\n",
      "Maximum file size: 3000000000B\n",
      "Adding sequences from FASTA; added 16350 sequences in 0.79114 seconds.\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Building a new DB, current time: 05/04/2025 19:05:12\n",
      "New DB name:   /research/projects/chlamydomonas/NIexpression/genome_info/Cschloesseri/longest_isoform_transcript.fa\n",
      "New DB title:  /research/projects/chlamydomonas/NIexpression/genome_info/Cschloesseri/longest_isoform_transcript.fa\n",
      "Sequence type: Nucleotide\n",
      "Deleted existing Nucleotide BLAST database named /research/projects/chlamydomonas/NIexpression/genome_info/Cschloesseri/longest_isoform_transcript.fa\n",
      "Keep MBits: T\n",
      "Maximum file size: 3000000000B\n",
      "Adding sequences from FASTA; added 15571 sequences in 0.727543 seconds.\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Building a new DB, current time: 05/04/2025 19:05:13\n",
      "New DB name:   /research/projects/chlamydomonas/NIexpression/genome_info/Malaysian/Cmalay.v1.proteins.longest_isoform.fasta\n",
      "New DB title:  /research/projects/chlamydomonas/NIexpression/genome_info/Malaysian/Cmalay.v1.proteins.longest_isoform.fasta\n",
      "Sequence type: Protein\n",
      "Deleted existing Protein BLAST database named /research/projects/chlamydomonas/NIexpression/genome_info/Malaysian/Cmalay.v1.proteins.longest_isoform.fasta\n",
      "Keep MBits: T\n",
      "Maximum file size: 3000000000B\n",
      "Adding sequences from FASTA; added 17730 sequences in 0.628954 seconds.\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Building a new DB, current time: 05/04/2025 19:05:14\n",
      "New DB name:   /research/projects/chlamydomonas/NIexpression/genome_info/Creinhardtii_v6.1/CreinhardtiiCC_4532_707_v6.1.protein_primaryTranscriptOnly.fa\n",
      "New DB title:  /research/projects/chlamydomonas/NIexpression/genome_info/Creinhardtii_v6.1/CreinhardtiiCC_4532_707_v6.1.protein_primaryTranscriptOnly.fa\n",
      "Sequence type: Protein\n",
      "Deleted existing Protein BLAST database named /research/projects/chlamydomonas/NIexpression/genome_info/Creinhardtii_v6.1/CreinhardtiiCC_4532_707_v6.1.protein_primaryTranscriptOnly.fa\n",
      "Keep MBits: T\n",
      "Maximum file size: 3000000000B\n",
      "Adding sequences from FASTA; added 17693 sequences in 0.757357 seconds.\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Building a new DB, current time: 05/04/2025 19:05:15\n",
      "New DB name:   /research/projects/chlamydomonas/NIexpression/genome_info/JG5/JG5-1.v1.proteins.longest_isoform.fasta\n",
      "New DB title:  /research/projects/chlamydomonas/NIexpression/genome_info/JG5/JG5-1.v1.proteins.longest_isoform.fasta\n",
      "Sequence type: Protein\n",
      "Deleted existing Protein BLAST database named /research/projects/chlamydomonas/NIexpression/genome_info/JG5/JG5-1.v1.proteins.longest_isoform.fasta\n",
      "Keep MBits: T\n",
      "Maximum file size: 3000000000B\n",
      "Adding sequences from FASTA; added 19723 sequences in 0.697869 seconds.\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Building a new DB, current time: 05/04/2025 19:05:16\n",
      "New DB name:   /research/projects/chlamydomonas/NIexpression/genome_info/Cincerta/longest_isoform_protein.fa\n",
      "New DB title:  /research/projects/chlamydomonas/NIexpression/genome_info/Cincerta/longest_isoform_protein.fa\n",
      "Sequence type: Protein\n",
      "Deleted existing Protein BLAST database named /research/projects/chlamydomonas/NIexpression/genome_info/Cincerta/longest_isoform_protein.fa\n",
      "Keep MBits: T\n",
      "Maximum file size: 3000000000B\n",
      "Adding sequences from FASTA; added 16350 sequences in 0.586285 seconds.\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Building a new DB, current time: 05/04/2025 19:05:17\n",
      "New DB name:   /research/projects/chlamydomonas/NIexpression/genome_info/Cschloesseri/longest_isoform_protein.fa\n",
      "New DB title:  /research/projects/chlamydomonas/NIexpression/genome_info/Cschloesseri/longest_isoform_protein.fa\n",
      "Sequence type: Protein\n",
      "Deleted existing Protein BLAST database named /research/projects/chlamydomonas/NIexpression/genome_info/Cschloesseri/longest_isoform_protein.fa\n",
      "Keep MBits: T\n",
      "Maximum file size: 3000000000B\n",
      "Adding sequences from FASTA; added 15571 sequences in 0.594647 seconds.\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!makeblastdb -dbtype nucl -in /research/projects/chlamydomonas/NIexpression/genome_info/Creinhardtii_v6.1/CreinhardtiiCC_4532_707_v6.1.transcript_primaryTranscriptOnly.fa\n",
    "!makeblastdb -dbtype nucl -in /research/projects/chlamydomonas/NIexpression/genome_info/Cincerta/longest_isoform_transcript.fa\n",
    "!makeblastdb -dbtype nucl -in /research/projects/chlamydomonas/NIexpression/genome_info/Cschloesseri/longest_isoform_transcript.fa\n",
    "\n",
    "!makeblastdb -dbtype prot -in /research/projects/chlamydomonas/NIexpression/genome_info/Malaysian/Cmalay.v1.proteins.longest_isoform.fasta\n",
    "!makeblastdb -dbtype prot -in /research/projects/chlamydomonas/NIexpression/genome_info/Creinhardtii_v6.1/CreinhardtiiCC_4532_707_v6.1.protein_primaryTranscriptOnly.fa\n",
    "!makeblastdb -dbtype prot -in /research/projects/chlamydomonas/NIexpression/genome_info/JG5/JG5-1.v1.proteins.longest_isoform.fasta\n",
    "!makeblastdb -dbtype prot -in /research/projects/chlamydomonas/NIexpression/genome_info/Cincerta/longest_isoform_protein.fa\n",
    "!makeblastdb -dbtype prot -in /research/projects/chlamydomonas/NIexpression/genome_info/Cschloesseri/longest_isoform_protein.fa\n",
    "\n",
    "\n",
    "# #### CREN vs JG5-1 ####\n",
    "# !blastx -query /research/projects/chlamydomonas/NIexpression/genome_info/Creinhardtii_v6.1/CreinhardtiiCC_4532_707_v6.1.transcript_primaryTranscriptOnly.fa -db /research/projects/chlamydomonas/NIexpression/genome_info/JG5/JG5-1.v1.proteins.longest_isoform.fasta -num_threads 48 -evalue 1e-10 -out Cr_JG5.blastx.txt -max_hsps 1 -num_alignments 1 -outfmt \"6 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore qcovs qcovhsp\"\n",
    "# !blastx -query /research/projects/chlamydomonas/NIexpression/genome_info/JG5/longest_isoform_transcript.fa -db /research/projects/chlamydomonas/NIexpression/genome_info/Creinhardtii_v6.1/CreinhardtiiCC_4532_707_v6.1.protein_primaryTranscriptOnly.fa -num_threads 48 -evalue 1e-10 -out JG5_Cr.blastx.txt -max_hsps 1 -num_alignments 1 -outfmt \"6 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore qcovs qcovhsp\"\n",
    "# #### CREN vs CMAL ####\n",
    "# !blastx -query /research/projects/chlamydomonas/NIexpression/genome_info/Creinhardtii_v6.1/CreinhardtiiCC_4532_707_v6.1.transcript_primaryTranscriptOnly.fa -db /research/projects/chlamydomonas/NIexpression/genome_info/Malaysian/Cmalay.v1.proteins.longest_isoform.fasta -num_threads 48 -evalue 1e-10 -out Cr_Cmal.blastx.txt -max_hsps 1 -num_alignments 1 -outfmt \"6 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore qcovs qcovhsp\"\n",
    "# !blastx -query /research/projects/chlamydomonas/NIexpression/genome_info/Malaysian/longest_isoform_transcript.fa -db /research/projects/chlamydomonas/NIexpression/genome_info/Creinhardtii_v6.1/CreinhardtiiCC_4532_707_v6.1.protein_primaryTranscriptOnly.fa -num_threads 48 -evalue 1e-10 -out Cmal_Cr.blastx.txt -max_hsps 1 -num_alignments 1 -outfmt \"6 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore qcovs qcovhsp\"\n",
    "# #### CREN vs CINC ####\n",
    "# !blastx -query /research/projects/chlamydomonas/NIexpression/genome_info/Cincerta/longest_isoform_transcript.fa -db /research/projects/chlamydomonas/NIexpression/genome_info/Creinhardtii_v6.1/CreinhardtiiCC_4532_707_v6.1.protein_primaryTranscriptOnly.fa -num_threads 48 -evalue 1e-10 -out Cinc_Cr.blastx.txt -max_hsps 1 -num_alignments 1 -outfmt \"6 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore qcovs qcovhsp\"\n",
    "# !blastx -query /research/projects/chlamydomonas/NIexpression/genome_info/Creinhardtii_v6.1/CreinhardtiiCC_4532_707_v6.1.transcript_primaryTranscriptOnly.fa -db /research/projects/chlamydomonas/NIexpression/genome_info/Cincerta/longest_isoform_protein.fa -num_threads 48 -evalue 1e-10 -out Cr_Cinc.blastx.txt -max_hsps 1 -num_alignments 1 -outfmt \"6 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore qcovs qcovhsp\"\n",
    "# #### CREN vs CSCH ####\n",
    "# !blastx -query /research/projects/chlamydomonas/NIexpression/genome_info/Cschloesseri/longest_isoform_transcript.fa -db /research/projects/chlamydomonas/NIexpression/genome_info/Creinhardtii_v6.1/CreinhardtiiCC_4532_707_v6.1.protein_primaryTranscriptOnly.fa -num_threads 48 -evalue 1e-10 -out Csch_Cr.blastx.txt -max_hsps 1 -num_alignments 1 -outfmt \"6 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore qcovs qcovhsp\"\n",
    "# !blastx -query /research/projects/chlamydomonas/NIexpression/genome_info/Creinhardtii_v6.1/CreinhardtiiCC_4532_707_v6.1.transcript_primaryTranscriptOnly.fa -db /research/projects/chlamydomonas/NIexpression/genome_info/Cschloesseri/longest_isoform_protein.fa -num_threads 48 -evalue 1e-10 -out Cr_Csch.blastx.txt -max_hsps 1 -num_alignments 1 -outfmt \"6 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore qcovs qcovhsp\"\n",
    "\n",
    "# #### JG5-1 vs CMAL ####\n",
    "# !blastx -query /research/projects/chlamydomonas/NIexpression/genome_info/JG5/longest_isoform_transcript.fa -db /research/projects/chlamydomonas/NIexpression/genome_info/Malaysian/Cmalay.v1.proteins.longest_isoform.fasta -num_threads 48 -evalue 1e-10 -out JG5_Cmal.blastx.txt -max_hsps 1 -num_alignments 1 -outfmt \"6 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore qcovs qcovhsp\"\n",
    "# !blastx -query /research/projects/chlamydomonas/NIexpression/genome_info/Malaysian/longest_isoform_transcript.fa -db /research/projects/chlamydomonas/NIexpression/genome_info/JG5/JG5-1.v1.proteins.longest_isoform.fasta -num_threads 48 -evalue 1e-10 -out Cmal_JG5.blastx.txt -max_hsps 1 -num_alignments 1 -outfmt \"6 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore qcovs qcovhsp\"\n",
    "# #### JG5-1 vs CINC ####\n",
    "# !blastx -query /research/projects/chlamydomonas/NIexpression/genome_info/JG5/longest_isoform_transcript.fa -db /research/projects/chlamydomonas/NIexpression/genome_info/Cincerta/longest_isoform_protein.fa -num_threads 48 -evalue 1e-10 -out JG5_Cinc.blastx.txt -max_hsps 1 -num_alignments 1 -outfmt \"6 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore qcovs qcovhsp\"\n",
    "# !blastx -query /research/projects/chlamydomonas/NIexpression/genome_info/Cincerta/longest_isoform_transcript.fa -db /research/projects/chlamydomonas/NIexpression/genome_info/JG5/JG5-1.v1.proteins.longest_isoform.fasta -num_threads 48 -evalue 1e-10 -out Cinc_JG5.blastx.txt -max_hsps 1 -num_alignments 1 -outfmt \"6 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore qcovs qcovhsp\"\n",
    "# #### JG5-1 vs CSCH ####\n",
    "# !blastx -query /research/projects/chlamydomonas/NIexpression/genome_info/JG5/longest_isoform_transcript.fa -db /research/projects/chlamydomonas/NIexpression/genome_info/Cschloesseri/longest_isoform_protein.fa -num_threads 48 -evalue 1e-10 -out JG5_Csch.blastx.txt -max_hsps 1 -num_alignments 1 -outfmt \"6 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore qcovs qcovhsp\"\n",
    "# !blastx -query /research/projects/chlamydomonas/NIexpression/genome_info/Cschloesseri/longest_isoform_transcript.fa -db /research/projects/chlamydomonas/NIexpression/genome_info/JG5/JG5-1.v1.proteins.longest_isoform.fasta -num_threads 48 -evalue 1e-10 -out Csch_JG5.blastx.txt -max_hsps 1 -num_alignments 1 -outfmt \"6 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore qcovs qcovhsp\"\n",
    "\n",
    "# #### CMAL vs CINC ####\n",
    "# !blastx -query /research/projects/chlamydomonas/NIexpression/genome_info/Cincerta/longest_isoform_transcript.fa -db /research/projects/chlamydomonas/NIexpression/genome_info/Malaysian/Cmalay.v1.proteins.longest_isoform.fasta -num_threads 48 -evalue 1e-10 -out Cinc_Cmal.blastx.txt -max_hsps 1 -num_alignments 1 -outfmt \"6 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore qcovs qcovhsp\"\n",
    "# !blastx -query /research/projects/chlamydomonas/NIexpression/genome_info/Malaysian/longest_isoform_transcript.fa -db /research/projects/chlamydomonas/NIexpression/genome_info/Cincerta/longest_isoform_protein.fa -num_threads 48 -evalue 1e-10 -out Cmal_Cinc.blastx.txt -max_hsps 1 -num_alignments 1 -outfmt \"6 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore qcovs qcovhsp\"\n",
    "# #### CMAL vs CSCH ####\n",
    "# !blastx -query /research/projects/chlamydomonas/NIexpression/genome_info/Cschloesseri/longest_isoform_transcript.fa -db /research/projects/chlamydomonas/NIexpression/genome_info/Malaysian/Cmalay.v1.proteins.longest_isoform.fasta -num_threads 48 -evalue 1e-10 -out Csch_Cmal.blastx.txt -max_hsps 1 -num_alignments 1 -outfmt \"6 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore qcovs qcovhsp\"\n",
    "# !blastx -query /research/projects/chlamydomonas/NIexpression/genome_info/Malaysian/longest_isoform_transcript.fa -db /research/projects/chlamydomonas/NIexpression/genome_info/Cschloesseri/longest_isoform_protein.fa -num_threads 48 -evalue 1e-10 -out Cmal_Csch.blastx.txt -max_hsps 1 -num_alignments 1 -outfmt \"6 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore qcovs qcovhsp\"\n",
    "\n",
    "# #### CINC vs CSCH ####\n",
    "# !blastx -query /research/projects/chlamydomonas/NIexpression/genome_info/Cschloesseri/longest_isoform_transcript.fa -db /research/projects/chlamydomonas/NIexpression/genome_info/Cincerta/longest_isoform_protein.fa -num_threads 48 -evalue 1e-10 -out Csch_Cinc.blastx.txt -max_hsps 1 -num_alignments 1 -outfmt \"6 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore qcovs qcovhsp\"\n",
    "# !blastx -query /research/projects/chlamydomonas/NIexpression/genome_info/Cincerta/longest_isoform_transcript.fa -db /research/projects/chlamydomonas/NIexpression/genome_info/Cschloesseri/longest_isoform_protein.fa -num_threads 48 -evalue 1e-10 -out Cinc_Csch.blastx.txt -max_hsps 1 -num_alignments 1 -outfmt \"6 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore qcovs qcovhsp\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "falling-tract",
   "metadata": {},
   "outputs": [],
   "source": [
    "#### Finding gene duplicates ####\n",
    "!blastx -query /research/projects/chlamydomonas/NIexpression/genome_info/JG5/longest_isoform_transcript.fa -db /research/projects/chlamydomonas/NIexpression/genome_info/JG5/JG5-1.v1.proteins.longest_isoform.fasta -num_threads 48 -evalue 1e-10 -out JG5_JG5.blastx.txt -max_hsps 1 -num_alignments 5 -outfmt \"6 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore qcovs qcovhsp\"\n",
    "!blastx -query /research/projects/chlamydomonas/NIexpression/genome_info/Cschloesseri/longest_isoform_transcript.fa -db /research/projects/chlamydomonas/NIexpression/genome_info/Cschloesseri/longest_isoform_protein.fa -num_threads 48 -evalue 1e-10 -out Csch_Csch.blastx.txt -max_hsps 1 -num_alignments 5 -outfmt \"6 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore qcovs qcovhsp\"\n",
    "!blastx -query /research/projects/chlamydomonas/NIexpression/genome_info/Cincerta/longest_isoform_transcript.fa -db /research/projects/chlamydomonas/NIexpression/genome_info/Cincerta/longest_isoform_protein.fa -num_threads 48 -evalue 1e-10 -out Cinc_Cinc.blastx.txt -max_hsps 1 -num_alignments 5 -outfmt \"6 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore qcovs qcovhsp\"\n",
    "!blastx -query /research/projects/chlamydomonas/NIexpression/genome_info/Malaysian/longest_isoform_transcript.fa -db /research/projects/chlamydomonas/NIexpression/genome_info/Malaysian/Cmalay.v1.proteins.longest_isoform.fasta -num_threads 48 -evalue 1e-10 -out Cmal_Cmal.blastx.txt -max_hsps 1 -num_alignments 5 -outfmt \"6 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore qcovs qcovhsp\"\n",
    "!blastx -query /research/projects/chlamydomonas/NIexpression/genome_info/Creinhardtii_v6.1/CreinhardtiiCC_4532_707_v6.1.transcript_primaryTranscriptOnly.fa -db /research/projects/chlamydomonas/NIexpression/genome_info/Creinhardtii_v6.1/CreinhardtiiCC_4532_707_v6.1.protein_primaryTranscriptOnly.fa -num_threads 48 -evalue 1e-10 -out Cren_Cren.blastx.txt -max_hsps 1 -num_alignments 5 -outfmt \"6 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore qcovs qcovhsp\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "unsigned-brave",
   "metadata": {},
   "source": [
    "# Rob's Code"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "incomplete-adoption",
   "metadata": {},
   "source": [
    "## BLAST line class"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "addressed-intelligence",
   "metadata": {},
   "outputs": [],
   "source": [
    "class blast_line:\n",
    "    def __init__(self, l):\n",
    "        l = l.strip()\n",
    "        self.query = l.split()[0].split('|')[0].split('_4532')[0].split('.t')[0]\n",
    "        self.subject = l.split()[1].split('|')[0].split('_4532')[0].split('.t')[0]\n",
    "        self.percent_id = float(l.split()[2])\n",
    "        self.alignment_length = int(l.split()[3]) \n",
    "        self.mistmatches = int(l.split()[4])\n",
    "        self.gap_openings = int(l.split()[5]) \n",
    "        self.query_start = int(l.split()[6]) \n",
    "        self.query_end = int(l.split()[7]) \n",
    "        self.subject_start = int(l.split()[8]) \n",
    "        self.subject_end = int(l.split()[9])\n",
    "        self.e_value = float(l.split()[10])\n",
    "        self.bit_score = float(l.split()[11])\n",
    "        self.qcovs = float(l.split()[12])\n",
    "        self.qcovhsp = float(l.split()[13])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "latest-extraction",
   "metadata": {},
   "source": [
    "## Assess the similarity of pairs\n",
    "--- \n",
    "\n",
    " - Using the BLAST output we an look at the paired CDSs based on their \n",
    "     - qCov -Coverage - the qcoverage of the shorter CDS\n",
    "     - PID - Percent Identity - the % ID of the better alignment\n",
    "\n",
    "We can look at the relationship of the two by filtering the pairs on a series of combinations of PID and qCOV\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "coupled-smell",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "def similarity_pairs(pairs):\n",
    "    df = pd.DataFrame()\n",
    "    min_qcovs = [60]\n",
    "    df['min_qcov'] = min_qcovs\n",
    "    for min_pid in ([60]):\n",
    "        col = []\n",
    "        for min_qcov in min_qcovs:\n",
    "            filtered_pairs = []\n",
    "            for p in pairs:\n",
    "                r,i = p\n",
    "                r_b = Cr2Ci[r]\n",
    "                i_b = Ci2Cr[i]\n",
    "                pids  = (r_b.percent_id,i_b.percent_id)\n",
    "                qcovs = (r_b.qcovs,i_b.qcovs)\n",
    "                if max(pids) > min_pid and max(qcovs) > min_qcov:\n",
    "                    #print('PASS', r, i, max(pids), max(qcovs))\n",
    "                    filtered_pairs.append(p)\n",
    "            col.append(len(filtered_pairs))\n",
    "        df[str(min_pid)] = col\n",
    "    \n",
    "    return df, filtered_pairs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "stuck-thong",
   "metadata": {},
   "source": [
    "## Parse Reciprocal Best BLAST Hits (RBH) into pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "functioning-fleet",
   "metadata": {},
   "outputs": [],
   "source": [
    "files = [('Cinc_Cr.blastx.txt', 'Cr_Cinc.blastx.txt', 'CINC_CREN'), ('Cmal_Cr.blastx.txt', 'Cr_Cmal.blastx.txt', 'CMAL_CREN'), \n",
    " ('Cr_Csch.blastx.txt', 'Csch_Cr.blastx.txt', 'CREN_CSCH'), ('Cr_JG5.blastx.txt', 'JG5_Cr.blastx.txt', 'CREN_JG5-1'), \n",
    "         ('JG5_Cmal.blastx.txt', 'Cmal_JG5.blastx.txt', 'JG5-1_CMAL'), ('JG5_Cinc.blastx.txt', 'Cinc_JG5.blastx.txt', 'JG5-1_CINC'), \n",
    " ('JG5_Csch.blastx.txt', 'Csch_JG5.blastx.txt', 'JG5-1_CSCH'), ('Cmal_Cinc.blastx.txt', 'Cinc_Cmal.blastx.txt', 'CMAL_CINC'), \n",
    "        ('Cmal_Csch.blastx.txt', 'Csch_Cmal.blastx.txt', 'CMAL_CSCH'), ('Csch_Cinc.blastx.txt', 'Cinc_Csch.blastx.txt', 'CSCH_CINC')]\n",
    "\n",
    "for file in files:\n",
    "    Cr2Ci = {}\n",
    "    for l in open(\"/research/projects/chlamydomonas/NIexpression/data/orthologs/\" + file[0]):\n",
    "        b = blast_line(l)\n",
    "        Cr2Ci[b.query] = b\n",
    "    \n",
    "    Ci2Cr = {}\n",
    "    for l in open(\"/research/projects/chlamydomonas/NIexpression/data/orthologs/\" + file[1]):\n",
    "        b = blast_line(l)\n",
    "        Ci2Cr[b.query] = b\n",
    "    \n",
    "    # Check for hits\n",
    "    df_pairs = pd.DataFrame()\n",
    "    pairs = []\n",
    "    for r in Cr2Ci:\n",
    "        hit = Cr2Ci[r]\n",
    "        i = hit.subject\n",
    "    #     print(r, i)\n",
    "    #     print(Ci2Cr[i].subject)\n",
    "        if i in Ci2Cr and Ci2Cr[i].subject == r: # Pair\n",
    "            pairs.append([r,i])\n",
    "#     print(len(pairs))\n",
    "    \n",
    "    ## Save to dataframe\n",
    "    df_pairs = pd.DataFrame(pairs)\n",
    "    df_pairs.to_csv('/research/projects/chlamydomonas/NIexpression/data/orthologs/' + file[2], sep = '\\t', index = False, header = False)\n",
    "\n",
    "    ## Save filtered pairs to dataframe\n",
    "    filtered = pd.DataFrame(similarity_pairs(pairs)[1])\n",
    "    filtered.columns = file[2].split('_')\n",
    "    filtered.to_csv('/research/projects/chlamydomonas/NIexpression/data/orthologs/filteredpairs_' + file[2], sep = '\\t', index = False, header = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "lasting-tonight",
   "metadata": {},
   "source": [
    "## Combining RBH Orthologous Pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "angry-exhaust",
   "metadata": {},
   "outputs": [],
   "source": [
    "CREN_CINC = pd.read_csv('/research/projects/chlamydomonas/NIexpression/data/orthologs/CINC_CREN', delimiter = '\\t', names = ['CINC', 'CREN'])\n",
    "CREN_CSCH = pd.read_csv('/research/projects/chlamydomonas/NIexpression/data/orthologs/CREN_CSCH', delimiter = '\\t', names = ['CREN', 'CSCH'])\n",
    "CREN_MALA = pd.read_csv('/research/projects/chlamydomonas/NIexpression/data/orthologs/CMAL_CREN', delimiter = '\\t', names = ['MALA', 'CREN'])\n",
    "CREN_JG05 = pd.read_csv('/research/projects/chlamydomonas/NIexpression/data/orthologs/CREN_JG5-1', delimiter = '\\t', names = ['CREN', 'JG05'])\n",
    "MALA_JG05 = pd.read_csv('/research/projects/chlamydomonas/NIexpression/data/orthologs/JG5-1_CMAL', delimiter = '\\t', names = ['JG05', 'MALA'])\n",
    "CINC_MALA = pd.read_csv('/research/projects/chlamydomonas/NIexpression/data/orthologs/CMAL_CINC', delimiter = '\\t', names = ['MALA', 'CINC'])\n",
    "CINC_CSCH = pd.read_csv('/research/projects/chlamydomonas/NIexpression/data/orthologs/CSCH_CINC', delimiter = '\\t', names = ['CSCH', 'CINC'])\n",
    "CSCH_MALA = pd.read_csv('/research/projects/chlamydomonas/NIexpression/data/orthologs/CMAL_CSCH', delimiter = '\\t', names = ['MALA', 'CSCH'])\n",
    "JG05_CINC = pd.read_csv('/research/projects/chlamydomonas/NIexpression/data/orthologs/JG5-1_CINC', delimiter = '\\t', names = ['JG05', 'CINC'])\n",
    "JG05_CSCH = pd.read_csv('/research/projects/chlamydomonas/NIexpression/data/orthologs/JG5-1_CSCH', delimiter = '\\t', names = ['JG05', 'CSCH'])\n",
    "\n",
    "result = pd.concat([CREN_CINC, CREN_CSCH, CREN_MALA, CREN_JG05, MALA_JG05, CINC_MALA, CINC_CSCH, CSCH_MALA, JG05_CINC, JG05_CSCH], axis = 0)\n",
    "result.to_csv('/research/projects/chlamydomonas/NIexpression/data/orthologs/BLASTRBH_orthologous_pairs.txt', sep = '\\t', index = False, header = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "negative-journal",
   "metadata": {},
   "source": [
    "## Combining Filtered RBH Orthologous Pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "mathematical-complexity",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "directory = '/research/projects/chlamydomonas/NIexpression/data/orthologs'\n",
    "filtered_orths = pd.DataFrame()\n",
    "    \n",
    "for filename in os.listdir(directory):\n",
    "    if re.findall('filteredpairs', filename):\n",
    "        f = os.path.join(directory, filename)\n",
    "        data = pd.read_csv(f, delimiter = '\\t')\n",
    "        filtered_orths = pd.concat([filtered_orths, data], axis = 0)\n",
    "species = ['CINC', 'CMAL', 'CSCH', 'JG5-1']\n",
    "for i in species:\n",
    "    filtered_orths[i] = filtered_orths[i].str.split('.t', expand = True)[0]\n",
    "filtered_orths.to_csv('/research/projects/chlamydomonas/NIexpression/data/orthologs/filteredRBH_orthologous_pairs.txt', sep = '\\t', index = False, header = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "automated-spice",
   "metadata": {},
   "source": [
    "## Sequence Divergence of Orthologous pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "wired-customs",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "directory = '/research/projects/chlamydomonas/InterspeciesCodingDivergence'\n",
    "divergence = pd.DataFrame()\n",
    "    \n",
    "for filename in os.listdir(directory):\n",
    "    if re.findall('.out.txt', filename):\n",
    "        f = os.path.join(directory, filename)\n",
    "        data = pd.read_csv(f, delimiter = '\\t')\n",
    "        data.columns = filename[:-8].split('_') + list(data.columns[2:])\n",
    "        for i in data.columns[0:2]:\n",
    "            try:\n",
    "                data[i] = data[i].str.split('_4532', expand = True)[0]\n",
    "            except: continue\n",
    "            try:\n",
    "                data[i] = data[i].str.split('.t', expand = True)[0]\n",
    "            except: continue\n",
    "        data.to_csv('/research/projects/chlamydomonas/NIexpression/data/orthologs/divergence_' + filename, sep = '\\t', index = False, header = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "numeric-shareware",
   "metadata": {},
   "source": [
    "# Finding Mating Type using BLASTn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "musical-seeking",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "Building a new DB, current time: 12/23/2024 15:01:37\n",
      "New DB name:   /research/projects/chlamydomonas/NIexpression/genome_info/Malaysian/Cmalay.v1.genome_w_organelles.fasta\n",
      "New DB title:  /research/projects/chlamydomonas/NIexpression/genome_info/Malaysian/Cmalay.v1.genome_w_organelles.fasta\n",
      "Sequence type: Nucleotide\n",
      "Deleted existing Nucleotide BLAST database named /research/projects/chlamydomonas/NIexpression/genome_info/Malaysian/Cmalay.v1.genome_w_organelles.fasta\n",
      "Keep MBits: T\n",
      "Maximum file size: 3000000000B\n",
      "Adding sequences from FASTA; added 92 sequences in 1.17132 seconds.\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Building a new DB, current time: 12/23/2024 15:01:39\n",
      "New DB name:   /research/projects/chlamydomonas/NIexpression/genome_info/JG5/JG5-1.v1.genome_w_organelles.fasta\n",
      "New DB title:  /research/projects/chlamydomonas/NIexpression/genome_info/JG5/JG5-1.v1.genome_w_organelles.fasta\n",
      "Sequence type: Nucleotide\n",
      "Deleted existing Nucleotide BLAST database named /research/projects/chlamydomonas/NIexpression/genome_info/JG5/JG5-1.v1.genome_w_organelles.fasta\n",
      "Keep MBits: T\n",
      "Maximum file size: 3000000000B\n",
      "Adding sequences from FASTA; added 132 sequences in 1.13819 seconds.\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Building a new DB, current time: 12/23/2024 15:01:40\n",
      "New DB name:   /research/projects/chlamydomonas/Cincerta_deNovo/analysis/assembly_V3/final_assemblies_annotations/Chlamydomonas_incerta/Chlamydomonas_incerta.V3_5.with_organelles.fa\n",
      "New DB title:  /research/projects/chlamydomonas/Cincerta_deNovo/analysis/assembly_V3/final_assemblies_annotations/Chlamydomonas_incerta/Chlamydomonas_incerta.V3_5.with_organelles.fa\n",
      "Sequence type: Nucleotide\n",
      "Keep MBits: T\n",
      "Maximum file size: 3000000000B\n",
      "Adding sequences from FASTA; added 455 sequences in 1.37961 seconds.\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Building a new DB, current time: 12/23/2024 15:01:42\n",
      "New DB name:   /research/projects/chlamydomonas/Cincerta_deNovo/analysis/assembly_V3/final_assemblies_annotations/Chlamydomonas_schloesseri/Chlamydomonas_schloesseri.V1_5.with_organelles.fa\n",
      "New DB title:  /research/projects/chlamydomonas/Cincerta_deNovo/analysis/assembly_V3/final_assemblies_annotations/Chlamydomonas_schloesseri/Chlamydomonas_schloesseri.V1_5.with_organelles.fa\n",
      "Sequence type: Nucleotide\n",
      "Keep MBits: T\n",
      "Maximum file size: 3000000000B\n",
      "Adding sequences from FASTA; added 459 sequences in 1.37888 seconds.\n",
      "\n",
      "\n",
      "Warning: [blastn] Examining 5 or more matches is recommended\n",
      "Warning: [blastn] Examining 5 or more matches is recommended\n",
      "Warning: [blastn] Examining 5 or more matches is recommended\n",
      "Warning: [blastn] Examining 5 or more matches is recommended\n"
     ]
    }
   ],
   "source": [
    "mtplus = /research/references/chlamydomonas/5.3_chlamy_w_organelles_mt_minus/mtPlus_ref.chromosome_6.fasta\n",
    "mtminus = /research/references/chlamydomonas/5.3_chlamy_w_organelles_mt_minus/mtMinus_ref.chromosome_6_and_mtMinus.fasta\n",
    "\n",
    "!makeblastdb -dbtype nucl -in /research/projects/chlamydomonas/NIexpression/genome_info/Malaysian/Cmalay.v1.genome_w_organelles.fasta\n",
    "!makeblastdb -dbtype nucl -in /research/projects/chlamydomonas/NIexpression/genome_info/JG5/JG5-1.v1.genome_w_organelles.fasta\n",
    "!makeblastdb -dbtype nucl -in /research/projects/chlamydomonas/Cincerta_deNovo/analysis/assembly_V3/final_assemblies_annotations/Chlamydomonas_incerta/Chlamydomonas_incerta.V3_5.with_organelles.fa\n",
    "!makeblastdb -dbtype nucl -in /research/projects/chlamydomonas/Cincerta_deNovo/analysis/assembly_V3/final_assemblies_annotations/Chlamydomonas_schloesseri/Chlamydomonas_schloesseri.V1_5.with_organelles.fa\n",
    "\n",
    "#### Malaysian ####\n",
    "!blastn -query /research/references/chlamydomonas/5.3_chlamy_w_organelles_mt_minus/mtPlus_ref.chromosome_6.fasta -db /research/projects/chlamydomonas/NIexpression/genome_info/Malaysian/Cmalay.v1.genome_w_organelles.fasta -num_threads 48 -evalue 1e-10 -out Cmal_v_mt+.blastn.txt -max_hsps 1 -num_alignments 1 -outfmt \"6 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore qcovs qcovhsp\"\n",
    "!blastn -query /research/references/chlamydomonas/5.3_chlamy_w_organelles_mt_minus/mtMinus_ref.chromosome_6_and_mtMinus.fasta -db /research/projects/chlamydomonas/NIexpression/genome_info/Malaysian/Cmalay.v1.genome_w_organelles.fasta -num_threads 48 -evalue 1e-10 -out Cmal_v_mt-.blastn.txt -max_hsps 1 -num_alignments 1 -outfmt \"6 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore qcovs qcovhsp\"\n",
    "#### JG5-1 ####\n",
    "!blastn -query /research/references/chlamydomonas/5.3_chlamy_w_organelles_mt_minus/mtPlus_ref.chromosome_6.fasta -db /research/projects/chlamydomonas/NIexpression/genome_info/JG5/JG5-1.v1.genome_w_organelles.fasta -num_threads 48 -evalue 1e-10 -out mating_type/JG5_v_mt+.blastn.txt -max_hsps 1 -num_alignments 1 -outfmt \"6 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore qcovs qcovhsp\"\n",
    "!blastn -query /research/references/chlamydomonas/5.3_chlamy_w_organelles_mt_minus/mtMinus_ref.chromosome_6_and_mtMinus.fasta -db /research/projects/chlamydomonas/NIexpression/genome_info/JG5/JG5-1.v1.genome_w_organelles.fasta -num_threads 48 -evalue 1e-10 -out mating_type/JG5_v_mt-.blastn.txt -max_hsps 1 -num_alignments 1 -outfmt \"6 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore qcovs qcovhsp\"\n",
    "#### Chlamydomonas incerta ####\n",
    "!blastn -query /research/references/chlamydomonas/5.3_chlamy_w_organelles_mt_minus/mtPlus_ref.chromosome_6.fasta -db /research/projects/chlamydomonas/Cincerta_deNovo/analysis/assembly_V3/final_assemblies_annotations/Chlamydomonas_incerta/Chlamydomonas_incerta.V3_5.with_organelles.fa -num_threads 48 -evalue 1e-10 -out mating_type/Cinc_v_mt+.blastn.txt -max_hsps 1 -num_alignments 1 -outfmt \"6 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore qcovs qcovhsp\"\n",
    "!blastn -query /research/references/chlamydomonas/5.3_chlamy_w_organelles_mt_minus/mtMinus_ref.chromosome_6_and_mtMinus.fasta -db /research/projects/chlamydomonas/Cincerta_deNovo/analysis/assembly_V3/final_assemblies_annotations/Chlamydomonas_incerta/Chlamydomonas_incerta.V3_5.with_organelles.fa -num_threads 48 -evalue 1e-10 -out mating_type/Cinc_v_mt-.blastn.txt -max_hsps 1 -num_alignments 1 -outfmt \"6 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore qcovs qcovhsp\"\n",
    "#### Chlamydomonas schloesseri ####\n",
    "!blastn -query /research/references/chlamydomonas/5.3_chlamy_w_organelles_mt_minus/mtPlus_ref.chromosome_6.fasta -db /research/projects/chlamydomonas/Cincerta_deNovo/analysis/assembly_V3/final_assemblies_annotations/Chlamydomonas_schloesseri/Chlamydomonas_schloesseri.V1_5.with_organelles.fa -num_threads 48 -evalue 1e-10 -out mating_type/Csch_v_mt+.blastn.txt -max_hsps 1 -num_alignments 1 -outfmt \"6 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore qcovs qcovhsp\"\n",
    "!blastn -query /research/references/chlamydomonas/5.3_chlamy_w_organelles_mt_minus/mtMinus_ref.chromosome_6_and_mtMinus.fasta -db /research/projects/chlamydomonas/Cincerta_deNovo/analysis/assembly_V3/final_assemblies_annotations/Chlamydomonas_schloesseri/Chlamydomonas_schloesseri.V1_5.with_organelles.fa -num_threads 48 -evalue 1e-10 -out mating_type/Csch_v_mt-.blastn.txt -max_hsps 1 -num_alignments 1 -outfmt \"6 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore qcovs qcovhsp\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "collective-gasoline",
   "metadata": {},
   "outputs": [],
   "source": [
    "CREN_CINC = pd.read_csv('/research/projects/chlamydomonas/NIexpression/data/orthologs/CREN_CINC', delimiter = '\\t', names = ['CINC', 'CREN'])\n",
    "CREN_CSCH = pd.read_csv('/research/projects/chlamydomonas/NIexpression/data/orthologs/CREN_CSCH', delimiter = '\\t', names = ['CREN', 'CSCH'])\n",
    "CREN_MALA = pd.read_csv('/research/projects/chlamydomonas/NIexpression/data/orthologs/CREN_CMAL', delimiter = '\\t', names = ['MALA', 'CREN'])\n",
    "CREN_JG05 = pd.read_csv('/research/projects/chlamydomonas/NIexpression/data/orthologs/CREN_JG5-1', delimiter = '\\t', names = ['CREN', 'JG05'])\n",
    "\n",
    "CREN_CINC['CINC'] = CREN_CINC['CINC'].str.split('.t', expand = True)[0]\n",
    "CREN_CSCH['CSCH'] = CREN_CSCH['CSCH'].str.split('.t', expand = True)[0]\n",
    "CREN_MALA['MALA'] = CREN_MALA['MALA'].str.split('.t', expand = True)[0]\n",
    "CREN_JG05['JG05'] = CREN_JG05['JG05'].str.split('.t', expand = True)[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "vital-judge",
   "metadata": {},
   "outputs": [],
   "source": [
    "a = pd.read_csv('/research/projects/chlamydomonas/NIexpression/data/raw_counts/CREN_orthologous_genes.csv', delimiter = '\\t', index_col = 'Unnamed: 0')\n",
    "\n",
    "crci = a[['CREN', 'CINC']].dropna()\n",
    "crcs = a[['CREN', 'CSCH']].dropna()\n",
    "crcm = a[['CREN', 'MALA']].dropna()\n",
    "crjg = a[['CREN', 'JG05']].dropna()\n",
    "\n",
    "crcs2 = pd.merge(crcs, CREN_CSCH, on = ['CREN', 'CSCH']).set_index('CREN')\n",
    "crcm2 = pd.merge(crcm, CREN_MALA, on = ['CREN', 'MALA']).set_index('CREN')\n",
    "crjg2 = pd.merge(crjg, CREN_JG05, on = ['CREN', 'JG05']).set_index('CREN')\n",
    "crci2 = pd.merge(crci, CREN_CINC, on = ['CREN', 'CINC']).set_index('CREN')\n",
    "\n",
    "a = crcs2.join(crcm2).join(crjg2).join(crci2).reset_index()\n",
    "\n",
    "a.to_csv('/research/projects/chlamydomonas/NIexpression/data/orthologs/CREN_orthologous_genes.csv', sep = '\\t', header = True, index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "generous-disney",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:.conda-personal] *",
   "language": "python",
   "name": "conda-env-.conda-personal-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.16"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": true,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
