{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "86c5b3a3-abf7-4565-8c71-7aff562361a1",
   "metadata": {},
   "source": [
    "Hi Rob, can you please get me the genetic distances of paralogous pairs found in my project folder: ~NIexpression/data/gene_duplicates/paralogs.csv? Thanks!\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "7a5bbd39-42ae-456b-bd30-9aa6b975157e",
   "metadata": {},
   "outputs": [],
   "source": [
    "from annotation import Transcript, GFF_line\n",
    "from Bio import SeqIO, AlignIO\n",
    "import glob, re, itertools, gzip\n",
    "\n",
    "\n",
    "from Bio.Align.Applications import MuscleCommandline\n",
    "from ness_fasta import ness_fasta\n",
    "import signal, gzip\n",
    "from Bio import AlignIO\n",
    "from Bio.Seq import Seq\n",
    "import subprocess\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "\n",
    "def signal_handler(signum, frame):\n",
    "    raise Exception(\"Timed out!\")\n",
    "from Bio import SeqIO\n",
    "import sys, subprocess, os, glob, re, Bio, random\n",
    "from Bio import AlignIO, SeqIO\n",
    "from Bio.Align import AlignInfo\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from collections import defaultdict\n",
    "#from Bio.Alphabet import IUPAC, Gapped\n",
    "#from Bio.Align.Generic import Alignment\n",
    "from Bio.Align import Alignment\n",
    "from Bio.Align import MultipleSeqAlignment\n",
    "#alphabet = Gapped(IUPAC.ambiguous_dna)\n",
    "\n",
    "from math import log\n",
    "from glob import glob\n",
    "from Bio import SeqIO, AlignIO, codonalign\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.Align import MultipleSeqAlignment\n",
    "\n",
    "\n",
    "def translation_aligner(seqfile, prefix, write=True):\n",
    "    #Pass fasta file\n",
    "    #Select the desired NCBI translation table, 1 = standard (http://www.ncbi.nlm.nih.gov/Taxonomy/Utils/wprintgc.cgi)\n",
    "    translationTable = 1\n",
    "    # take list of sequences and read the fasta sequences into a dictionary\n",
    "    seq_dict = SeqIO.to_dict(SeqIO.parse(open(seqfile), 'fasta'))\n",
    "    # Translate the sequences\n",
    "    aa_seqs = []\n",
    "    for key in seq_dict:\n",
    "        aaSeq = SeqRecord(seq_dict[key].seq.translate(table=translationTable), id=key)\n",
    "        aa_seqs.append(aaSeq)\n",
    "    # Replace stop codons with X (unknown aa) so muscle doesn't drop them\n",
    "    for aaSeq in aa_seqs:\n",
    "        noStopCodonSeq = str(aaSeq.seq).replace('*', 'X')\n",
    "        aaSeq.seq = Seq(noStopCodonSeq)\n",
    "    # write aa seq to file\n",
    "    o = open(\"tmp.aa.unALN.fasta\", 'w')\n",
    "    SeqIO.write(aa_seqs, o, 'fasta')\n",
    "    o.close()\n",
    "    # do the alignment\n",
    "    cmd = \"muscle -align tmp.aa.unALN.fasta -output tmp.aa.ALN.fasta\"  \n",
    "    child = subprocess.Popen(cmd, stdout=subprocess.PIPE, stderr = subprocess.PIPE, shell=True) #don't pipe stderr or muscle hangs\n",
    "    stdout, stderr = child.communicate()\n",
    "    aaAlignment = AlignIO.read(\"tmp.aa.ALN.fasta\", 'fasta')\n",
    "    #Convert the aa alignment into a dna alignment\n",
    "    dnaAlignment = MultipleSeqAlignment([])\n",
    "    for aa_seq in aaAlignment:\n",
    "        aaCount = 0\n",
    "        dnaSeq = ''\n",
    "        for aa in aa_seq.seq:\n",
    "            if (aa == '-'):\n",
    "                dnaSeq +=  '---'\n",
    "            else:\n",
    "                dnaSeq += seq_dict[aa_seq.id].seq[aaCount*3:aaCount*3+3]\n",
    "                aaCount+=1\n",
    "        dnaAlignment.append(SeqRecord(Seq(str(dnaSeq)), id = aa_seq.id))\n",
    "    if write == True:\n",
    "        o = open(prefix + '.aln.fasta', 'w')\n",
    "        AlignIO.write([dnaAlignment], o, 'fasta')\n",
    "        o.close()\n",
    "    return dnaAlignment\n",
    "\n",
    "\n",
    "def degenerate(seq):\n",
    "    degen_seq = ''\n",
    "    for idx in range(0, len(seq)-2,3):    \n",
    "        codon = seq[idx:idx+3]\n",
    "        try:\n",
    "            degen  = degen_dict[codon.lower()]\n",
    "        except KeyError:\n",
    "            degen = 'nnn'\n",
    "        degen_seq = degen_seq+degen\n",
    "    return degen_seq\n",
    "\n",
    "\n",
    "def JCdistance(diffs, seq_length):\n",
    "    \"\"\" \n",
    "    distance = -b log(1 - p / b)\n",
    "    where:\n",
    "    b = 3/4\n",
    "    and p = p-distance, diffs/seq_length i.e. uncorrected distance between sequences \n",
    "    \"\"\"\n",
    "    b = 0.75\n",
    "    try: p = float(diffs)/seq_length\n",
    "    except ZeroDivisionError:\n",
    "        return None\n",
    "    try: K = -b * log( 1-p/b )\n",
    "    except ValueError: \n",
    "           return None\n",
    "    return K\n",
    "\n",
    "\n",
    "def k0k4(translation_alignment_file):\n",
    "    with open(translation_alignment_file,'r') as f:\n",
    "        nuc_aln = AlignIO.read(f, 'fasta')\n",
    "        nuc1 = nuc_aln[0]\n",
    "        nuc2 = nuc_aln[1]\n",
    "        degen_map = degenerate(str(nuc1.seq))\n",
    "        data = {'diffs':{'0':0,'2':0,'4':0}, 'sites':{'0':0,'2':0,'4':0}}\n",
    "        for idx in range(len(degen_map)):\n",
    "            d, base1, base2 = [thing[idx] for thing in [degen_map, str(nuc1.seq), str(nuc2.seq)]]\n",
    "            if base1 in 'ATGC' and base2 in 'ATGC' and d in '024':\n",
    "                data['sites'][d]+=1\n",
    "                if base1 != base2:\n",
    "                    data['diffs'][d]+=1\n",
    "        k0 = JCdistance(data['diffs']['0'], data['sites']['0'])\n",
    "        k4 = JCdistance(data['diffs']['4'], data['sites']['4'])\n",
    "        total_sites = sum([data['sites'][i] for i in data['sites']])\n",
    "        return(k0,k4,total_sites)\n",
    "\n",
    "        \n",
    "        \n",
    "degen_dict = {\n",
    "    'ttt': '002', 'ttc': '002', 'tta': '202', 'ttg': '202', \\\n",
    "    'tct': '004', 'tcc': '004', 'tca': '004', 'tcg': '004', \\\n",
    "    'tat': '002', 'tac': '002', 'taa': '022', 'tag': '002', \\\n",
    "    'tgt': '002', 'tgc': '002', 'tga': '020', 'tgg': '000', \\\n",
    "    'ctt': '004', 'ctc': '004', 'cta': '204', 'ctg': '204', \\\n",
    "    'cct': '004', 'ccc': '004', 'cca': '004', 'ccg': '004', \\\n",
    "    'cat': '002', 'cac': '002', 'caa': '002', 'cag': '002', \\\n",
    "    'cgt': '004', 'cgc': '004', 'cga': '204', 'cgg': '204', \\\n",
    "    'att': '003', 'atc': '003', 'ata': '003', 'atg': '000', \\\n",
    "    'act': '004', 'acc': '004', 'aca': '004', 'acg': '004', \\\n",
    "    'aat': '002', 'aac': '002', 'aaa': '002', 'aag': '002', \\\n",
    "    'agt': '002', 'agc': '002', 'aga': '202', 'agg': '202', \\\n",
    "    'gtt': '004', 'gtc': '004', 'gta': '004', 'gtg': '004', \\\n",
    "    'gct': '004', 'gcc': '004', 'gca': '004', 'gcg': '004', \\\n",
    "    'gat': '002', 'gac': '002', 'gaa': '002', 'gag': '002', \\\n",
    "    'ggt': '004', 'ggc': '004', 'gga': '004', 'ggg': '004'\n",
    "}\n",
    "\n",
    "codons = {\n",
    "    \"TTT\": \"F\", \"TTC\": \"F\", \"TTA\": \"L\", \"TTG\": \"L\",\n",
    "    \"TCT\": \"S\", \"TCC\": \"S\", \"TCA\": \"S\", \"TCG\": \"S\",\n",
    "    \"TAT\": \"Y\", \"TAC\": \"Y\", \"TAA\": \"*\", \"TAG\": \"*\",\n",
    "    \"TGT\": \"C\", \"TGC\": \"C\", \"TGA\": \"*\", \"TGG\": \"W\",\n",
    "    \"CTT\": \"L\", \"CTC\": \"L\", \"CTA\": \"L\", \"CTG\": \"L\",\n",
    "    \"CCT\": \"P\", \"CCC\": \"P\", \"CCA\": \"P\", \"CCG\": \"P\",\n",
    "    \"CAT\": \"H\", \"CAC\": \"H\", \"CAA\": \"Q\", \"CAG\": \"Q\",\n",
    "    \"CGT\": \"R\", \"CGC\": \"R\", \"CGA\": \"R\", \"CGG\": \"R\",\n",
    "    \"ATT\": \"I\", \"ATC\": \"I\", \"ATA\": \"I\", \"ATG\": \"M\",\n",
    "    \"ACT\": \"T\", \"ACC\": \"T\", \"ACA\": \"T\", \"ACG\": \"T\",\n",
    "    \"AAT\": \"N\", \"AAC\": \"N\", \"AAA\": \"K\", \"AAG\": \"K\",\n",
    "    \"AGT\": \"S\", \"AGC\": \"S\", \"AGA\": \"R\", \"AGG\": \"R\",\n",
    "    \"GTT\": \"V\", \"GTC\": \"V\", \"GTA\": \"V\", \"GTG\": \"V\",\n",
    "    \"GCT\": \"A\", \"GCC\": \"A\", \"GCA\": \"A\", \"GCG\": \"A\",\n",
    "    \"GAT\": \"D\", \"GAC\": \"D\", \"GAA\": \"E\", \"GAG\": \"E\",\n",
    "    \"GGT\": \"G\", \"GGC\": \"G\", \"GGA\": \"G\", \"GGG\": \"G\",\n",
    "    \"---\": \"-\"}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "71a173ee-3b09-4be3-afae-03e3049dd8a5",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "CINC_CDS = {}\n",
    "for s in SeqIO.parse('/research/projects/chlamydomonas/NIexpression/genome_info/Cincerta/longest_isoform_transcript.fa', 'fasta'):\n",
    "    k = s.id.split(\".\")[0]\n",
    "    CINC_CDS[k] = s\n",
    "    \n",
    "CSCH_CDS = {}\n",
    "for s in SeqIO.parse('/research/projects/chlamydomonas/NIexpression/genome_info/Cschloesseri/longest_isoform_transcript.fa', 'fasta'):\n",
    "    k = s.id.split(\".\")[0]\n",
    "    CSCH_CDS[k] = s\n",
    "    \n",
    "MALA_CDS = {}\n",
    "for s in SeqIO.parse('/research/projects/chlamydomonas/NIexpression/genome_info/Malaysian/longest_isoform_transcript.fa', 'fasta'):\n",
    "    k = s.id.split(\".\")[0]\n",
    "    MALA_CDS[k] = s\n",
    "\n",
    "\n",
    "JG05_CDS = {}\n",
    "for s in SeqIO.parse('/research/projects/chlamydomonas/NIexpression/genome_info/JG5/longest_isoform_transcript.fa', 'fasta'):\n",
    "    k = s.id.split(\".\")[0]\n",
    "    JG05_CDS[k] = s\n",
    "    \n",
    "CREN_CDS = {}\n",
    "for s in SeqIO.parse(gzip.open('/research/references/chlamydomonas/6.1_chlamy/annotation/CreinhardtiiCC_4532_707_v6.1.cds_primaryTranscriptOnly.fa.gz', 'rt'), 'fasta'):\n",
    "    k = s.id.split(\"_4532.1\")[0]\n",
    "    CREN_CDS[k] = s\n",
    "\n",
    "CDSs = {'CINC': CINC_CDS,\n",
    "        'CREN': CREN_CDS,\n",
    "        'CSCH': CSCH_CDS,\n",
    "        'CMAL': MALA_CDS,\n",
    "        'JG5': JG05_CDS,\n",
    "       }\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "7549ccc6-bf98-439a-b6c3-6ec98d5017ed",
   "metadata": {},
   "outputs": [],
   "source": [
    "paralog_files = glob.glob(\"../data/gene_duplicates/*_paralogousgenepairs*\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "5599f15f-1a05-4fcc-9f93-4048745e53e0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mkdir: cannot create directory ‘../data/gene_duplicates/translation_alignments/’: Permission denied\n"
     ]
    }
   ],
   "source": [
    "mkdir ../data/gene_duplicates/translation_alignments/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "34c1d6ff-51d6-4561-908a-649e8aeb19f1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CREN\n",
      "\t 0 297\n",
      "CINC\n",
      "\t 0 142\n",
      "CSCH\n",
      "\t 0 134\n",
      "CMAL\n",
      "\t 0 114\n",
      "JG5\n",
      "\t 0 169\n"
     ]
    }
   ],
   "source": [
    "bc = 0\n",
    "\n",
    "for paralog_file in paralog_files:\n",
    "    species = paralog_file.split(\"/\")[-1].split(\"_\")[0]\n",
    "    print(species)\n",
    "    missing = 0\n",
    "    total =0 \n",
    "    output = open(\"../data/gene_duplicates/{}.paralog_divergence.tsv\".format(species), 'w')\n",
    "    genes =  CDSs[species]\n",
    "    for line in open(paralog_file).readlines()[1:]:\n",
    "        pair = line.strip().split()\n",
    "        cds1, cds2 = genes[pair[0]], genes[pair[1]]\n",
    "        with open('../data/gene_duplicates/translation_alignments/tmp.unaligned.fasta','w') as o:\n",
    "            o.write(\">{}\\n{}\\n\".format(pair[0], cds1.seq)) \n",
    "            o.write(\">{}\\n{}\\n\".format(pair[1], cds2.seq))\n",
    "        try:\n",
    "            aligned = translation_aligner('../data/gene_duplicates/translation_alignments/tmp.unaligned.fasta', \"../data/gene_duplicates/translation_alignments/translation.tmp\".format(species), write=True)\n",
    "            total+=1\n",
    "        except Exception as msg:\n",
    "            print(\"\\tPair {} timed out!\\n\".format(pair) )\n",
    "            missing +=1\n",
    "            total+=1\n",
    "            continue\n",
    "        k0,k4, total_sites = k0k4('../data/gene_duplicates/translation_alignments/translation.tmp.aln.fasta')\n",
    "        data = [pair[0], pair[1], total_sites, k0, k4]\n",
    "        output.write(\"\\t\".join([str(i) for i in data]) + \"\\n\")\n",
    "        \n",
    "        if bc > 2:\n",
    "            pass\n",
    "        bc +=1\n",
    "    output.close()\n",
    "    print(\"\\t\", missing, total)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c4306151-0999-4255-81a1-5a4b292799b2",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.19"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": true,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
